{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Eitan177/Bioinformatics_Bootcamp/blob/main/informatics_bootcamp.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Welcome to the GOAL Bioinformatics Bootcamp, 2024! To Setup this notebook please run the first three cells. This will download the packages you will need to participate"
      ],
      "metadata": {
        "id": "fKTuyMLINOKx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# NGS Bioinformatics Hands On Lab\n",
        "\n",
        "# Install required packages\n",
        "!apt-get update\n",
        "!apt-get install -y fastp fastqc bwa samtools tabix\n",
        "!pip install pysam\n",
        "\n",
        "# Optional: Install additional tools if not available in Colab\n",
        "!wget https://github.com/biod/sambamba/releases/download/v0.8.0/sambamba-0.8.0-linux-amd64-static.gz\n",
        "!gunzip sambamba-0.8.0-linux-amd64-static.gz\n",
        "!chmod +x sambamba-0.8.0-linux-amd64-static\n",
        "!mv sambamba-0.8.0-linux-amd64-static /usr/local/bin/sambamba\n",
        "\n",
        "# For ABRA2\n",
        "!wget https://github.com/mozack/abra2/releases/download/v2.23/abra2-2.23.jar\n",
        "!chmod +x abra2-2.23.jar\n",
        "\n",
        "# For VarScan\n",
        "!wget https://github.com/dkoboldt/varscan/raw/master/VarScan.v2.4.6.jar\n",
        "!chmod +x VarScan.v2.4.6.jar\n",
        "\n",
        "# for ANNOVAR\n",
        "!wget http://www.openbioinformatics.org/annovar/download/0wgxR2rIVP/annovar.latest.tar.gz\n",
        "!tar -zxvf annovar.latest.tar.gz\n",
        "!rm -r annovar.latest.tar.gz\n",
        "!rm -r annovar/humandb\n",
        "!perl annovar/annotate_variation.pl -buildver hg19 -downdb -webfrom annovar refGene assets/humandb/\n",
        "!perl /content/annovar/annotate_variation.pl -buildver hg19 -downdb cytoBand assets/humandb/\n",
        "!perl annovar/annotate_variation.pl -buildver hg19 -downdb -webfrom annovar clinvar_20150330 assets/humandb/\n",
        "!perl annovar/annotate_variation.pl -buildver hg19 -downdb -webfrom annovar cosmic85 assets/humandb/\n",
        "\n",
        "# Create directories for data and reference files\n",
        "!mkdir -p data assets/gatk_b37 assets/humandb annovar\n",
        "\n",
        "# Download pre-indexed reference genome\n",
        "!wget https://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/analysisSet/hg19.p13.plusMT.no_alt_analysis_set.bwa_index.tar.gz -O assets/hg19_bwa_index.tar.gz\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IkUOHAp-FmAu",
        "outputId": "1391e2b9-8422-4c3d-dcb5-7bea9f90423a"
      },
      "execution_count": 74,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rHit:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease\n",
            "\r0% [Connecting to archive.ubuntu.com] [Connecting to security.ubuntu.com (185.125.190.83)] [Connecte\r                                                                                                    \rHit:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease\n",
            "Hit:3 https://r2u.stat.illinois.edu/ubuntu jammy InRelease\n",
            "Hit:4 http://security.ubuntu.com/ubuntu jammy-security InRelease\n",
            "Hit:5 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Hit:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease\n",
            "Hit:7 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Hit:8 http://archive.ubuntu.com/ubuntu jammy-backports InRelease\n",
            "Hit:9 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "bwa is already the newest version (0.7.17-6).\n",
            "fastp is already the newest version (0.20.1+dfsg-1).\n",
            "fastqc is already the newest version (0.11.9+dfsg-5).\n",
            "samtools is already the newest version (1.13-4).\n",
            "tabix is already the newest version (1.13+ds-2build1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 30 not upgraded.\n",
            "Requirement already satisfied: pysam in /usr/local/lib/python3.11/dist-packages (0.23.0)\n",
            "--2025-03-20 21:42:06--  https://github.com/biod/sambamba/releases/download/v0.8.0/sambamba-0.8.0-linux-amd64-static.gz\n",
            "Resolving github.com (github.com)... 140.82.114.4\n",
            "Connecting to github.com (github.com)|140.82.114.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/4167482/7e411b00-3323-11eb-9622-b063c6aa7fc6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250320%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250320T214207Z&X-Amz-Expires=300&X-Amz-Signature=c28eb3be4ac3cf50a8ab821731966d221ed3345fd085e85b29e18ab3df50206e&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dsambamba-0.8.0-linux-amd64-static.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2025-03-20 21:42:07--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/4167482/7e411b00-3323-11eb-9622-b063c6aa7fc6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250320%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250320T214207Z&X-Amz-Expires=300&X-Amz-Signature=c28eb3be4ac3cf50a8ab821731966d221ed3345fd085e85b29e18ab3df50206e&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dsambamba-0.8.0-linux-amd64-static.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.108.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.111.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 4715787 (4.5M) [application/octet-stream]\n",
            "Saving to: ‘sambamba-0.8.0-linux-amd64-static.gz’\n",
            "\n",
            "sambamba-0.8.0-linu 100%[===================>]   4.50M  --.-KB/s    in 0.09s   \n",
            "\n",
            "2025-03-20 21:42:07 (51.0 MB/s) - ‘sambamba-0.8.0-linux-amd64-static.gz’ saved [4715787/4715787]\n",
            "\n",
            "--2025-03-20 21:42:07--  https://github.com/mozack/abra2/releases/download/v2.23/abra2-2.23.jar\n",
            "Resolving github.com (github.com)... 140.82.113.4\n",
            "Connecting to github.com (github.com)|140.82.113.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/66582878/e5e1b580-f6a6-11ea-89c8-33c01403f120?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250320%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250320T214208Z&X-Amz-Expires=300&X-Amz-Signature=5eff12823211fa0dd33cbf85e61efd5dd84288d6d232121c45ed5e4b63400982&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dabra2-2.23.jar&response-content-type=application%2Foctet-stream [following]\n",
            "--2025-03-20 21:42:08--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/66582878/e5e1b580-f6a6-11ea-89c8-33c01403f120?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250320%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250320T214208Z&X-Amz-Expires=300&X-Amz-Signature=5eff12823211fa0dd33cbf85e61efd5dd84288d6d232121c45ed5e4b63400982&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dabra2-2.23.jar&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.111.133, 185.199.109.133, 185.199.108.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.111.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 9025192 (8.6M) [application/octet-stream]\n",
            "Saving to: ‘abra2-2.23.jar.1’\n",
            "\n",
            "abra2-2.23.jar.1    100%[===================>]   8.61M  --.-KB/s    in 0.1s    \n",
            "\n",
            "2025-03-20 21:42:08 (80.4 MB/s) - ‘abra2-2.23.jar.1’ saved [9025192/9025192]\n",
            "\n",
            "--2025-03-20 21:42:08--  https://github.com/dkoboldt/varscan/raw/master/VarScan.v2.4.6.jar\n",
            "Resolving github.com (github.com)... 140.82.114.4\n",
            "Connecting to github.com (github.com)|140.82.114.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://raw.githubusercontent.com/dkoboldt/varscan/master/VarScan.v2.4.6.jar [following]\n",
            "--2025-03-20 21:42:08--  https://raw.githubusercontent.com/dkoboldt/varscan/master/VarScan.v2.4.6.jar\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 126611 (124K) [application/octet-stream]\n",
            "Saving to: ‘VarScan.v2.4.6.jar.1’\n",
            "\n",
            "VarScan.v2.4.6.jar. 100%[===================>] 123.64K  --.-KB/s    in 0.02s   \n",
            "\n",
            "2025-03-20 21:42:08 (4.86 MB/s) - ‘VarScan.v2.4.6.jar.1’ saved [126611/126611]\n",
            "\n",
            "--2025-03-20 21:42:09--  http://www.openbioinformatics.org/annovar/download/0wgxR2rIVP/annovar.latest.tar.gz\n",
            "Resolving www.openbioinformatics.org (www.openbioinformatics.org)... 67.205.156.247\n",
            "Connecting to www.openbioinformatics.org (www.openbioinformatics.org)|67.205.156.247|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 280727629 (268M) [application/x-gzip]\n",
            "Saving to: ‘annovar.latest.tar.gz’\n",
            "\n",
            "annovar.latest.tar. 100%[===================>] 267.72M  46.8MB/s    in 6.4s    \n",
            "\n",
            "2025-03-20 21:42:15 (41.7 MB/s) - ‘annovar.latest.tar.gz’ saved [280727629/280727629]\n",
            "\n",
            "annovar/\n",
            "annovar/example/\n",
            "annovar/example/ex1.avinput\n",
            "annovar/example/example.simple_region\n",
            "annovar/example/example.tab_region\n",
            "annovar/example/ex2.vcf\n",
            "annovar/example/grantham.matrix\n",
            "annovar/example/snplist.txt\n",
            "annovar/example/README\n",
            "annovar/example/gene_xref.txt\n",
            "annovar/example/gene_fullxref.txt\n",
            "annovar/humandb/\n",
            "annovar/humandb/hg19_refGene.txt\n",
            "annovar/humandb/hg19_refGeneMrna.fa\n",
            "annovar/humandb/hg19_refGeneVersion.txt\n",
            "annovar/humandb/hg19_refGeneWithVer.txt\n",
            "annovar/humandb/hg19_refGeneWithVerMrna.fa\n",
            "annovar/humandb/hg38_ensGeneMrna.fa\n",
            "annovar/humandb/hg38_ensGene.txt\n",
            "annovar/humandb/hg38_refGeneWithVer.txt\n",
            "annovar/humandb/hg38_refGeneWithVerMrna.fa\n",
            "annovar/humandb/hg19_example_db_generic.txt\n",
            "annovar/humandb/hg19_example_db_gff3.txt\n",
            "annovar/humandb/GRCh37_MT_ensGene.txt\n",
            "annovar/humandb/GRCh37_MT_ensGeneMrna.fa\n",
            "annovar/humandb/hg19_MT_ensGene.txt\n",
            "annovar/humandb/hg19_MT_ensGeneMrna.fa\n",
            "annovar/humandb/genometrax-sample-files-gff/\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_mirna_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_microsatellites_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_hgmd_disease_genes_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_evs_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_pathway_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_tcga_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_vista_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_cpg_islands_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_transfac_sites_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_dbnsfp_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_omim_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/list\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_dnase_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_hgmd_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_dbscsnv_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_clinvar_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_fulldbsnp_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_complete_genomics_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_tss_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_ptms_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_orpha_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_ethnicsnp_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_chip_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_drug_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_hgmdimputed_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_pgmd_featuretype_hg38.gff\n",
            "annovar/annotate_variation.pl\n",
            "annovar/table_annovar.pl\n",
            "annovar/convert2annovar.pl\n",
            "annovar/retrieve_seq_from_fasta.pl\n",
            "annovar/coding_change.pl\n",
            "annovar/variants_reduction.pl\n",
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_refGene.txt.gz ... OK\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_refGeneMrna.fa.gz ... OK\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_refGeneVersion.txt.gz ... OK\n",
            "NOTICE: Uncompressing downloaded files\n",
            "NOTICE: Finished downloading annotation files for hg19 build version, with files saved at the 'assets/humandb' directory\n",
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/cytoBand.txt.gz ... OK\n",
            "NOTICE: Uncompressing downloaded files\n",
            "NOTICE: Finished downloading annotation files for hg19 build version, with files saved at the 'assets/humandb' directory\n",
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_clinvar_20150330.txt.gz ... OK\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_clinvar_20150330.txt.idx.gz ... OK\n",
            "NOTICE: Uncompressing downloaded files\n",
            "NOTICE: Finished downloading annotation files for hg19 build version, with files saved at the 'assets/humandb' directory\n",
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_cosmic85.txt.gz ... Failed\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_cosmic85.txt.idx.gz ... Failed\n",
            "WARNING: Some files cannot be downloaded, including http://www.openbioinformatics.org/annovar/download/hg19_cosmic85.txt.idx.gz, http://www.openbioinformatics.org/annovar/download/hg19_cosmic85.txt.gz\n",
            "--2025-03-20 21:42:44--  https://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/analysisSet/hg19.p13.plusMT.no_alt_analysis_set.bwa_index.tar.gz\n",
            "Resolving hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)|128.114.119.163|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 3443961889 (3.2G) [application/x-gzip]\n",
            "Saving to: ‘assets/hg19_bwa_index.tar.gz’\n",
            "\n",
            "assets/hg19_bwa_ind 100%[===================>]   3.21G  56.0MB/s    in 60s     \n",
            "\n",
            "2025-03-20 21:43:45 (54.4 MB/s) - ‘assets/hg19_bwa_index.tar.gz’ saved [3443961889/3443961889]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/analysisSet/hg19.p13.plusMT.no_alt_analysis_set.fa.gz -O assets/hg19fa_index.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PCLBrAfFSDo0",
        "outputId": "ddad56db-d556-41a5-c4a2-1bf38617afab"
      },
      "execution_count": 75,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-03-20 21:44:15--  https://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/analysisSet/hg19.p13.plusMT.no_alt_analysis_set.fa.gz\n",
            "Resolving hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)|128.114.119.163|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 863259336 (823M) [application/x-gzip]\n",
            "Saving to: ‘assets/hg19fa_index.tar.gz’\n",
            "\n",
            "assets/hg19fa_index 100%[===================>] 823.27M  55.2MB/s    in 20s     \n",
            "\n",
            "2025-03-20 21:44:35 (42.0 MB/s) - ‘assets/hg19fa_index.tar.gz’ saved [863259336/863259336]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/analysisSet/hg19.p13.plusMT.no_alt_analysis_set.fa.gz -O assets/hg19fa_index.gz\n",
        "!gunzip assets/hg19fa_index.gz -C assets/gatk_b37/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QM7Of6eXSS1g",
        "outputId": "d04f445d-c266-45aa-d78f-e507f3284284"
      },
      "execution_count": 76,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-03-20 21:44:41--  https://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/analysisSet/hg19.p13.plusMT.no_alt_analysis_set.fa.gz\n",
            "Resolving hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)|128.114.119.163|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 863259336 (823M) [application/x-gzip]\n",
            "Saving to: ‘assets/hg19fa_index.gz’\n",
            "\n",
            "assets/hg19fa_index 100%[===================>] 823.27M  26.3MB/s    in 21s     \n",
            "\n",
            "2025-03-20 21:45:02 (39.9 MB/s) - ‘assets/hg19fa_index.gz’ saved [863259336/863259336]\n",
            "\n",
            "gzip: invalid option -- 'C'\n",
            "Try `gzip --help' for more information.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gunzip /content/assets/hg19fa_index.gz -c > assets/gatk_b37/hg19.fa"
      ],
      "metadata": {
        "id": "wEA2jlqJStHZ"
      },
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# 1. Examine content of the FASTQ file\n",
        "!zcat /content/ngs_ws_hd701_R1.fq.gz | head -n 20\n",
        "\n",
        "# 2. Perform FASTQ processing prior to alignment\n",
        "!fastp -i /content/ngs_ws_hd701_R1.fq.gz -o data/hd701_R1_processed.fq.gz -I /content/ngs_ws_hd701_R2.fq.gz -O data/hd701_R2_processed.fq.gz -z 4 -w 2 -h data/hd701_fq_qc.html\n",
        "\n",
        "# 3. Generate FASTQ quality control data\n",
        "!fastqc -o data -f fastq /content/ngs_ws_hd701_R1.fq.gz /content/ngs_ws_hd701_R2.fq.gz\n",
        "\n",
        "# 4. Display QC report in Colab\n",
        "from IPython.display import HTML\n",
        "with open('data/hd701_fq_qc.html', 'r') as f:\n",
        "    html_content = f.read()\n",
        "HTML(html_content)\n"
      ],
      "metadata": {
        "id": "4hZqzfN5J0r5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# 5. Align sequence to GRCh37 - Update the BWA reference path\n",
        "!bwa mem -M -v 1 -t 2 -R \"@RG\\\\tID:hd701\\\\tSM:hd701\\\\tPL:ILLUMINA\\\\tPI:150\\\\tCN:lab\" /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.p13.plusMT.no_alt_analysis_set.fa data/hd701_R1_processed.fq.gz data/hd701_R2_processed.fq.gz > data/hd701_raw.sam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DNiIZprhTEfh",
        "outputId": "9537dea6-113f-4b48-89de-7ddff22caa4b"
      },
      "execution_count": 77,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (144, 177, 224)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 384)\n",
            "[M::mem_pestat] mean and std.dev: (186.75, 58.21)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 464)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem -M -v 1 -t 2 -R @RG\\tID:hd701\\tSM:hd701\\tPL:ILLUMINA\\tPI:150\\tCN:lab /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.p13.plusMT.no_alt_analysis_set.fa data/hd701_R1_processed.fq.gz data/hd701_R2_processed.fq.gz\n",
            "[main] Real time: 40.311 sec; CPU: 7.902 sec\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 6. Convert the SAM file into a BAM file\n",
        "!samtools view -Shu data/hd701_raw.sam > data/hd701_raw.bam\n",
        "\n",
        "# 7. Sort and index raw BAM file\n",
        "!sambamba sort -p -t 2 -o data/hd701_sorted.bam data/hd701_raw.bam\n",
        "\n",
        "# 8. Mark PCR duplicates in BAM file\n",
        "!sambamba markdup -t 2 -p data/hd701_sorted.bam data/hd701_dedup.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sHwuMotpQ0BN",
        "outputId": "2805d5c1-5945-4480-a6be-52471787cdf9"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "sambamba 0.8.0\n",
            " by Artem Tarasov and Pjotr Prins (C) 2012-2020\n",
            "    LDC 1.10.0 / DMD v2.080.1 / LLVM6.0.1 / bootstrap LDC - the LLVM D compiler (0.17.4)\n",
            "\n",
            "Writing sorted chunks to temporary directory...\n",
            "[==============================================================================]\n",
            "\n",
            "sambamba 0.8.0\n",
            " by Artem Tarasov and Pjotr Prins (C) 2012-2020\n",
            "    LDC 1.10.0 / DMD v2.080.1 / LLVM6.0.1 / bootstrap LDC - the LLVM D compiler (0.17.4)\n",
            "\n",
            "finding positions of the duplicate reads in the file...\n",
            "[==============================================================================]\n",
            "  sorted 4959 end pairs\n",
            "     and 0 single ends (among them 0 unmatched pairs)\n",
            "  collecting indices of duplicate reads...   done in 2 ms\n",
            "  found 860 duplicates\n",
            "collected list of positions in 0 min 0 sec\n",
            "marking duplicates...\n",
            "[==============================================================================]\n",
            "collected list of positions in 0 min 0 sec\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 9. Perform indel realignment - Update the reference path\n",
        "!java -jar abra2-2.23.jar --in data/hd701_dedup.bam --out data/hd701_realigned.bam --ref /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa --threads 2 --targets assets/ngs_ws.bed --index --tmpdir /tmp/ > data/hd701_abra.log\n",
        "\n",
        "# 10. Inspect aligned sequences (show first few lines)\n",
        "!samtools view -h data/hd701_realigned.bam | head -n 20\n",
        "\n"
      ],
      "metadata": {
        "id": "TPom2kN_QAYl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools index data/hd701_realigned.bam\n",
        "!ls -l /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa\n",
        "!head -n 5 /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa\n",
        "!ls -l assets/ngs_ws.bed\n",
        "!head -n 5 assets/ngs_ws.bed\n",
        "!java -version\n",
        "!ls -l VarScan.v2.4.6.jar\n",
        "!head -n 10 data/hd701_mpileup.txt\n",
        "!wc -l data/hd701_mpileup.txt"
      ],
      "metadata": {
        "id": "6lhXsb1VVkjQ"
      },
      "execution_count": 27,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools mpileup -BA -q 20 -Q 30 -d 4000 -l assets/ngs_ws.bed -f /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa data/hd701_realigned.bam > data/hd701_mpileup.txt\n",
        "!wc -l data/hd701_mpileup.txt\n",
        "# Step 2: Run VarScan on the mpileup file\n",
        "#!cat data/hd701_mpileup.txt | java -Xmx4G -jar VarScan.v2.4.6.jar mpileup2vcf --min-coverage 8 --min-var-freq 0.05 --p-value 0.05 --min-avg-qual 30 --strand-filter 1 --output-vcf 1 --variants > data/hd701_raw.vcf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nZHuJyLiV3I4",
        "outputId": "87fb36d1-7adf-41cb-ad32-b065b0805357"
      },
      "execution_count": 36,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[mpileup] 1 samples in 1 input files\n",
            "8078 data/hd701_mpileup.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!# 11. Call variant - Update the reference path\n",
        "!samtools mpileup -BA -q 20 -Q 30 -d 4000 -l assets/ngs_ws.bed -f /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa data/hd701_realigned.bam | java -Xmx4G -jar VarScan.v2.4.6.jar mpileup2vcf --min-coverage 8 --min-var-freq 0.05 --p-value 0.05 --min-avg-qual 30 --strand-filter 1 --output-vcf 1 --variants > data/hd701_raw.vcf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "vSBVTyONUxOF",
        "outputId": "74d3213b-be53-45af-d9e6-9a9b8dece57f"
      },
      "execution_count": 38,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[mpileup] 1 samples in 1 input files\n",
            "Only variants will be reported\n",
            "Min coverage:\t8\n",
            "Min reads2:\t2\n",
            "Min var freq:\t0.05\n",
            "Min avg qual:\t30\n",
            "P-value thresh:\t0.05\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Reading input from STDIN\n",
            "8078 bases in pileup file\n",
            "14 variant positions (10 SNP, 4 indel)\n",
            "0 were failed by the strand-filter\n",
            "14 variant positions reported (10 SNP, 4 indel)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 12. Compress and index VCF file\n",
        "!bgzip data/hd701_raw.vcf\n",
        "!tabix -p vcf data/hd701_raw.vcf.gz\n",
        "\n",
        "# 13. Inspect variants in a raw VCF file\n",
        "!zcat data/hd701_raw.vcf.gz | head -n 20\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "10pXsS4ZXjUB",
        "outputId": "cca3a977-2a5c-481d-bd75-0f9d6faed360"
      },
      "execution_count": 39,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bgzip] data/hd701_raw.vcf.gz already exists; do you wish to overwrite (y or n)? y\n",
            "##fileformat=VCFv4.3\n",
            "##source=VarScan2\n",
            "##INFO=<ID=ADP,Number=1,Type=Integer,Description=\"Average per-sample depth of bases with Phred score >= 30\">\n",
            "##INFO=<ID=WT,Number=1,Type=Integer,Description=\"Number of samples called reference (wild-type)\">\n",
            "##INFO=<ID=HET,Number=1,Type=Integer,Description=\"Number of samples called heterozygous-variant\">\n",
            "##INFO=<ID=HOM,Number=1,Type=Integer,Description=\"Number of samples called homozygous-variant\">\n",
            "##INFO=<ID=NC,Number=1,Type=Integer,Description=\"Number of samples not called\">\n",
            "##FILTER=<ID=str10,Description=\"Less than 10% or more than 90% of variant supporting reads on one strand\">\n",
            "##FILTER=<ID=indelError,Description=\"Likely artifact due to indel reads at this position\">\n",
            "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
            "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
            "##FORMAT=<ID=SDP,Number=1,Type=Integer,Description=\"Raw Read Depth as reported by SAMtools\">\n",
            "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Quality Read Depth of bases with Phred score >= 30\">\n",
            "##FORMAT=<ID=RD,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases (reads1)\">\n",
            "##FORMAT=<ID=AD,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases (reads2)\">\n",
            "##FORMAT=<ID=FREQ,Number=1,Type=String,Description=\"Variant allele frequency\">\n",
            "##FORMAT=<ID=PVAL,Number=1,Type=String,Description=\"P-value from Fisher's Exact Test\">\n",
            "##FORMAT=<ID=RBQ,Number=1,Type=Integer,Description=\"Average quality of reference-supporting bases (qual1)\">\n",
            "##FORMAT=<ID=ABQ,Number=1,Type=Integer,Description=\"Average quality of variant-supporting bases (qual2)\">\n",
            "##FORMAT=<ID=RDF,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on forward strand (reads1plus)\">\n",
            "NOTICE: the --polish argument is set ON automatically (use --nopolish to change this behavior)\n",
            "Error: the required database file assets/humandb/hg19_refGene.txt does not exist.\n",
            "head: cannot open 'data/hd701_annotated.hg19_multianno.vcf' for reading: No such file or directory\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6dxKn1fxYOIe",
        "outputId": "d680940e-2bc7-4ed0-dd2a-0f9333a01a47"
      },
      "execution_count": 64,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/cytoBand.txt.gz ... OK\n",
            "NOTICE: Uncompressing downloaded files\n",
            "NOTICE: Finished downloading annotation files for hg19 build version, with files saved at the 'assets/humandb' directory\n",
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/hg19_cosmic85.txt.txt.gz ... Failed\n",
            "WARNING: Some files cannot be downloaded, including http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/hg19_cosmic85.txt.txt.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!perl /content/annovar/annotate_variation.pl -buildver hg19 -downdb hg19_cosmic assets/humandb/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5cpJuPPdcOLK",
        "outputId": "2179dd5d-a12d-4954-981b-1014fd973993"
      },
      "execution_count": 67,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/hg19_cosmic.txt.gz ... Failed\n",
            "WARNING: Some files cannot be downloaded, including http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/hg19_cosmic.txt.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget  https://ftp.ncbi.nlm.nih.gov/pub/ANNOVAR/hg19_cosmic85.txt.gz\n",
        "!wget  https://ftp.ncbi.nlm.nih.gov/pub/ANNOVAR/hg19_cosmic85.txt.idx.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tVXz0a1NdB7c",
        "outputId": "701cc9db-3361-49da-9dd3-5283db98fad2"
      },
      "execution_count": 73,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-03-20 21:30:43--  https://ftp.ncbi.nlm.nih.gov/pub/ANNOVAR/hg19_cosmic85.txt.gz\n",
            "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.11, 130.14.250.12, 130.14.250.13, ...\n",
            "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.11|:443... connected.\n",
            "HTTP request sent, awaiting response... 404 Not Found\n",
            "2025-03-20 21:30:43 ERROR 404: Not Found.\n",
            "\n",
            "--2025-03-20 21:30:43--  https://ftp.ncbi.nlm.nih.gov/pub/ANNOVAR/hg19_cosmic85.txt.idx.gz\n",
            "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.11, 130.14.250.12, 130.14.250.13, ...\n",
            "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.11|:443... connected.\n",
            "HTTP request sent, awaiting response... 404 Not Found\n",
            "2025-03-20 21:30:43 ERROR 404: Not Found.\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Gw3ifCzBYy0k",
        "outputId": "2361216f-c285-4c3d-f587-86405865cb39"
      },
      "execution_count": 51,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.gz ... Failed\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.idx.gz ... Failed\n",
            "WARNING: Some files cannot be downloaded, including http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.gz, http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.idx.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 14. Annotate variants (requires ANNOVAR to be set up)\n",
        "!perl annovar/table_annovar.pl data/hd701_raw.vcf.gz assets/humandb/ -out data/hd701_annotated -buildver hg19 -remove -nastring . -otherinfo -vcfinput -thread 2 -maxgenethread 2 -protocol refGene,cytoBand,cosmic85,clinvar_20150330 -operation g,r,f,f\n",
        "\n",
        "# 15. Inspect annotated VCF file\n",
        "!head -n 20 data/hd701_annotated.hg19_multianno.vcf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MeRm3HGFXqbe",
        "outputId": "91a8b835-1fd8-49d1-c811-52585b6c4aca"
      },
      "execution_count": 62,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "NOTICE: the --polish argument is set ON automatically (use --nopolish to change this behavior)\n",
            "Error: the required database file assets/humandb/hg19_cosmic85.txt does not exist.\n",
            "head: cannot open 'data/hd701_annotated.hg19_multianno.vcf' for reading: No such file or directory\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 11. Call variant - Update the reference path\n",
        "!samtools mpileup -BA -q 20 -Q 30 -d 4000 -l assets/ngs_ws.bed -f /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa data/hd701_realigned.bam | java -Xmx4G -jar VarScan.v2.4.6.jar mpileup2vcf --min-coverage 8 --min-var-freq 0.05 --p-value 0.05 --min-avg-qual 30 --strand-filter 1 --output-vcf 1 --variants > data/hd701_raw.vcf\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "# 16. Visualize BAM files with IGV.js\n",
        "!pip install igv-jupyter\n",
        "import igv\n",
        "viewer = igv.Browser({\n",
        "    \"genome\": \"hg19\",\n",
        "    \"locus\": \"chr17:7,571,720-7,590,868\"  # TP53 region\n",
        "})\n",
        "viewer.load_track({\n",
        "    \"name\": \"HD701 Alignment\",\n",
        "    \"url\": \"data/hd701_realigned.bam\",\n",
        "    \"indexURL\": \"data/hd701_realigned.bam.bai\",\n",
        "    \"format\": \"bam\"\n",
        "})\n",
        "viewer\n",
        "\n",
        "# Download results to your local machine\n",
        "from google.colab import files\n",
        "!zip -r hd701_results.zip data/\n",
        "files.download('hd701_results.zip')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "6aB5zjpZQSsI",
        "outputId": "bd94a03c-834c-430d-b200-0e6f64ac91f4"
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Only variants will be reported\n",
            "Min coverage:\t8\n",
            "Min reads2:\t2\n",
            "Min var freq:\t0.05\n",
            "Min avg qual:\t30\n",
            "P-value thresh:\t0.05\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "[bgzip] data/hd701_raw.vcf.gz already exists; do you wish to overwrite (y or n)? ^C\n",
            "[tabix] the index file exists. Please use '-f' to overwrite.\n",
            "NOTICE: the --polish argument is set ON automatically (use --nopolish to change this behavior)\n",
            "Error: the required database file assets/humandb/hg19_refGene.txt does not exist.\n",
            "head: cannot open 'data/hd701_annotated.hg19_multianno.vcf' for reading: No such file or directory\n",
            "Requirement already satisfied: igv-jupyter in /usr/local/lib/python3.11/dist-packages (2.0.1)\n",
            "Requirement already satisfied: jupyterlab~=3.0 in /usr/local/lib/python3.11/dist-packages (from igv-jupyter) (3.6.8)\n",
            "Requirement already satisfied: ipywidgets>=7.0.0 in /usr/local/lib/python3.11/dist-packages (from igv-jupyter) (7.7.1)\n",
            "Requirement already satisfied: igv-notebook in /usr/local/lib/python3.11/dist-packages (from igv-jupyter) (0.6.2)\n",
            "Requirement already satisfied: ipykernel>=4.5.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (6.17.1)\n",
            "Requirement already satisfied: ipython-genutils~=0.2.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (0.2.0)\n",
            "Requirement already satisfied: traitlets>=4.3.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (5.7.1)\n",
            "Requirement already satisfied: widgetsnbextension~=3.6.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (3.6.10)\n",
            "Requirement already satisfied: ipython>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (7.34.0)\n",
            "Requirement already satisfied: jupyterlab-widgets>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (3.0.13)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (24.2)\n",
            "Requirement already satisfied: tornado>=6.1.0 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (6.4.2)\n",
            "Requirement already satisfied: jupyter-core in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (5.7.2)\n",
            "Requirement already satisfied: jupyterlab-server~=2.19 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (2.27.3)\n",
            "Requirement already satisfied: jupyter-server<3,>=1.16.0 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (2.15.0)\n",
            "Requirement already satisfied: jupyter-ydoc~=0.2.4 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (0.2.5)\n",
            "Requirement already satisfied: jupyter-server-ydoc~=0.8.0 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (0.8.0)\n",
            "Requirement already satisfied: nbclassic in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (1.2.0)\n",
            "Requirement already satisfied: notebook<7 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (6.5.7)\n",
            "Requirement already satisfied: jinja2>=2.1 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (3.1.6)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from igv-notebook->igv-jupyter) (2.32.3)\n",
            "Requirement already satisfied: debugpy>=1.0 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (1.8.0)\n",
            "Requirement already satisfied: jupyter-client>=6.1.12 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (7.4.9)\n",
            "Requirement already satisfied: matplotlib-inline>=0.1 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (0.1.7)\n",
            "Requirement already satisfied: nest-asyncio in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (1.6.0)\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (5.9.5)\n",
            "Requirement already satisfied: pyzmq>=17 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (24.0.1)\n",
            "Requirement already satisfied: setuptools>=18.5 in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (75.1.0)\n",
            "Requirement already satisfied: jedi>=0.16 in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.19.2)\n",
            "Requirement already satisfied: decorator in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (4.4.2)\n",
            "Requirement already satisfied: pickleshare in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.7.5)\n",
            "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (3.0.50)\n",
            "Requirement already satisfied: pygments in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (2.18.0)\n",
            "Requirement already satisfied: backcall in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.2.0)\n",
            "Requirement already satisfied: pexpect>4.3 in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (4.9.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2>=2.1->jupyterlab~=3.0->igv-jupyter) (3.0.2)\n",
            "Requirement already satisfied: anyio>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (4.9.0)\n",
            "Requirement already satisfied: argon2-cffi>=21.1 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (23.1.0)\n",
            "Requirement already satisfied: jupyter-events>=0.11.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.12.0)\n",
            "Requirement already satisfied: jupyter-server-terminals>=0.4.4 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.5.3)\n",
            "Requirement already satisfied: nbconvert>=6.4.4 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (7.16.6)\n",
            "Requirement already satisfied: nbformat>=5.3.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (5.10.4)\n",
            "Requirement already satisfied: overrides>=5.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (7.7.0)\n",
            "Requirement already satisfied: prometheus-client>=0.9 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.21.1)\n",
            "Requirement already satisfied: send2trash>=1.8.2 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.8.3)\n",
            "Requirement already satisfied: terminado>=0.8.3 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.18.1)\n",
            "Requirement already satisfied: websocket-client>=1.7 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.8.0)\n",
            "Requirement already satisfied: platformdirs>=2.5 in /usr/local/lib/python3.11/dist-packages (from jupyter-core->jupyterlab~=3.0->igv-jupyter) (4.3.6)\n",
            "Requirement already satisfied: jupyter-server-fileid<1,>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server-ydoc~=0.8.0->jupyterlab~=3.0->igv-jupyter) (0.9.3)\n",
            "Requirement already satisfied: ypy-websocket<0.9.0,>=0.8.2 in /usr/local/lib/python3.11/dist-packages (from jupyter-server-ydoc~=0.8.0->jupyterlab~=3.0->igv-jupyter) (0.8.4)\n",
            "Requirement already satisfied: y-py<0.7.0,>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-ydoc~=0.2.4->jupyterlab~=3.0->igv-jupyter) (0.6.2)\n",
            "Requirement already satisfied: babel>=2.10 in /usr/local/lib/python3.11/dist-packages (from jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (2.17.0)\n",
            "Requirement already satisfied: json5>=0.9.0 in /usr/local/lib/python3.11/dist-packages (from jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (0.10.0)\n",
            "Requirement already satisfied: jsonschema>=4.18.0 in /usr/local/lib/python3.11/dist-packages (from jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (4.23.0)\n",
            "Requirement already satisfied: notebook-shim>=0.2.3 in /usr/local/lib/python3.11/dist-packages (from nbclassic->jupyterlab~=3.0->igv-jupyter) (0.2.4)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->igv-notebook->igv-jupyter) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests->igv-notebook->igv-jupyter) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->igv-notebook->igv-jupyter) (2.3.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests->igv-notebook->igv-jupyter) (2025.1.31)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.11/dist-packages (from anyio>=3.1.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.3.1)\n",
            "Requirement already satisfied: typing_extensions>=4.5 in /usr/local/lib/python3.11/dist-packages (from anyio>=3.1.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (4.12.2)\n",
            "Requirement already satisfied: argon2-cffi-bindings in /usr/local/lib/python3.11/dist-packages (from argon2-cffi>=21.1->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (21.2.0)\n",
            "Requirement already satisfied: parso<0.9.0,>=0.8.4 in /usr/local/lib/python3.11/dist-packages (from jedi>=0.16->ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.8.4)\n",
            "Requirement already satisfied: attrs>=22.2.0 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (25.3.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (2024.10.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (0.36.2)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (0.23.1)\n",
            "Requirement already satisfied: entrypoints in /usr/local/lib/python3.11/dist-packages (from jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (0.4)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (2.8.2)\n",
            "Requirement already satisfied: python-json-logger>=2.0.4 in /usr/local/lib/python3.11/dist-packages (from jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (3.3.0)\n",
            "Requirement already satisfied: pyyaml>=5.3 in /usr/local/lib/python3.11/dist-packages (from jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (6.0.2)\n",
            "Requirement already satisfied: rfc3339-validator in /usr/local/lib/python3.11/dist-packages (from jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.1.4)\n",
            "Requirement already satisfied: rfc3986-validator>=0.1.1 in /usr/local/lib/python3.11/dist-packages (from jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.1.1)\n",
            "Requirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (4.13.3)\n",
            "Requirement already satisfied: bleach!=5.0.0 in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (6.2.0)\n",
            "Requirement already satisfied: defusedxml in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.7.1)\n",
            "Requirement already satisfied: jupyterlab-pygments in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.3.0)\n",
            "Requirement already satisfied: mistune<4,>=2.0.3 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (3.1.2)\n",
            "Requirement already satisfied: nbclient>=0.5.0 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.10.2)\n",
            "Requirement already satisfied: pandocfilters>=1.4.1 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.5.1)\n",
            "Requirement already satisfied: fastjsonschema>=2.15 in /usr/local/lib/python3.11/dist-packages (from nbformat>=5.3.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (2.21.1)\n",
            "Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.11/dist-packages (from pexpect>4.3->ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.7.0)\n",
            "Requirement already satisfied: wcwidth in /usr/local/lib/python3.11/dist-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.2.13)\n",
            "Requirement already satisfied: aiofiles<23,>=22.1.0 in /usr/local/lib/python3.11/dist-packages (from ypy-websocket<0.9.0,>=0.8.2->jupyter-server-ydoc~=0.8.0->jupyterlab~=3.0->igv-jupyter) (22.1.0)\n",
            "Requirement already satisfied: aiosqlite<1,>=0.17.0 in /usr/local/lib/python3.11/dist-packages (from ypy-websocket<0.9.0,>=0.8.2->jupyter-server-ydoc~=0.8.0->jupyterlab~=3.0->igv-jupyter) (0.21.0)\n",
            "Requirement already satisfied: webencodings in /usr/local/lib/python3.11/dist-packages (from bleach!=5.0.0->bleach[css]!=5.0.0->nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.5.1)\n",
            "Requirement already satisfied: tinycss2<1.5,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.4.0)\n",
            "Requirement already satisfied: fqdn in /usr/local/lib/python3.11/dist-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.5.1)\n",
            "Requirement already satisfied: isoduration in /usr/local/lib/python3.11/dist-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (20.11.0)\n",
            "Requirement already satisfied: jsonpointer>1.13 in /usr/local/lib/python3.11/dist-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (3.0.0)\n",
            "Requirement already satisfied: uri-template in /usr/local/lib/python3.11/dist-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.3.0)\n",
            "Requirement already satisfied: webcolors>=24.6.0 in /usr/local/lib/python3.11/dist-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (24.11.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (1.17.0)\n",
            "Requirement already satisfied: cffi>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from argon2-cffi-bindings->argon2-cffi>=21.1->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.17.1)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4->nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (2.6)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.11/dist-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi>=21.1->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (2.22)\n",
            "Requirement already satisfied: arrow>=0.15.0 in /usr/local/lib/python3.11/dist-packages (from isoduration->jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.3.0)\n",
            "Requirement already satisfied: types-python-dateutil>=2.8.10 in /usr/local/lib/python3.11/dist-packages (from arrow>=0.15.0->isoduration->jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (2.9.0.20241206)\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "AttributeError",
          "evalue": "module 'igv' has no attribute 'Browser'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-24-76971d50519f>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'pip install igv-jupyter'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0migv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m viewer = igv.Browser({\n\u001b[0m\u001b[1;32m     22\u001b[0m     \u001b[0;34m\"genome\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m\"hg19\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m     \u001b[0;34m\"locus\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m\"chr17:7,571,720-7,590,868\"\u001b[0m  \u001b[0;31m# TP53 region\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mAttributeError\u001b[0m: module 'igv' has no attribute 'Browser'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!fastqc -o data -f fastq /content/ngs_ws_hd701_R1.fq.gz /content/ngs_ws_hd701_R2.fq.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Kuvbnln8OA_n",
        "outputId": "15c35715-2af8-4210-d15f-571e54a0c646"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of ngs_ws_hd701_R1.fq.gz\n",
            "Approx 20% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 40% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 60% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 80% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 100% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Analysis complete for ngs_ws_hd701_R1.fq.gz\n",
            "Started analysis of ngs_ws_hd701_R2.fq.gz\n",
            "Approx 20% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 40% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 60% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 80% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 100% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Analysis complete for ngs_ws_hd701_R2.fq.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import subprocess\n",
        "\n",
        "# Define input and output paths\n",
        "fastq_r1 = '/content/ngs_ws_hd701_R1.fq.gz'\n",
        "fastq_r2 = '/content/ngs_ws_hd701_R2.fq.gz'\n",
        "ref_genome = 'assets/gatk_b37/human_g1k_v37.fasta'\n",
        "output_prefix = 'data/ngs_ws_hd701'\n",
        "threads = 4  # Adjust based on your environment\n",
        "\n",
        "# Step 1: Quality control with FastQC (optional but recommended)\n",
        "print(\"Running FastQC...\")\n",
        "os.makedirs('data/fastqc', exist_ok=True)\n",
        "subprocess.run(['fastqc', fastq_r1, fastq_r2, '-o', 'data/fastqc'])\n",
        "\n",
        "# Step 2: Quality trimming with fastp\n",
        "print(\"Running fastp for quality trimming...\")\n",
        "subprocess.run([\n",
        "    'fastp',\n",
        "    '-i', fastq_r1,\n",
        "    '-I', fastq_r2,\n",
        "    '-o', f'{output_prefix}_trimmed_R1.fastq.gz',\n",
        "    '-O', f'{output_prefix}_trimmed_R2.fastq.gz',\n",
        "    '--json', f'{output_prefix}_fastp.json',\n",
        "    '--html', f'{output_prefix}_fastp.html'\n",
        "])\n",
        "\n",
        "# Step 3: Align to reference with BWA-MEM\n",
        "print(\"Aligning to reference genome...\")\n",
        "sam_file = f'{output_prefix}.sam'\n",
        "bwa_cmd = [\n",
        "    'bwa', 'mem',\n",
        "    '-t', str(threads),\n",
        "    '-R', f'@RG\\\\tID:ngs_ws_hd701\\\\tSM:ngs_ws_hd701\\\\tPL:ILLUMINA',\n",
        "    ref_genome,\n",
        "    f'{output_prefix}_trimmed_R1.fastq.gz',\n",
        "    f'{output_prefix}_trimmed_R2.fastq.gz'\n",
        "]\n",
        "\n",
        "with open(sam_file, 'w') as f:\n",
        "    subprocess.run(bwa_cmd, stdout=f)\n",
        "\n",
        "# Step 4: Convert SAM to BAM\n",
        "print(\"Converting SAM to BAM...\")\n",
        "bam_file = f'{output_prefix}.bam'\n",
        "subprocess.run(['samtools', 'view', '-b', '-h', '-o', bam_file, sam_file])\n",
        "\n",
        "# Step 5: Sort the BAM file\n",
        "print(\"Sorting BAM...\")\n",
        "sorted_bam = f'{output_prefix}.sorted.bam'\n",
        "subprocess.run(['samtools', 'sort', '-o', sorted_bam, bam_file])\n",
        "\n",
        "# Step 6: Index the BAM file\n",
        "print(\"Indexing BAM...\")\n",
        "subprocess.run(['samtools', 'index', sorted_bam])\n",
        "\n",
        "# Step 7: Mark duplicates with sambamba\n",
        "print(\"Marking duplicates...\")\n",
        "dedup_bam = f'{output_prefix}.dedup.bam'\n",
        "subprocess.run([\n",
        "    'sambamba', 'markdup',\n",
        "    '--tmpdir=/tmp',\n",
        "    '-t', str(threads),\n",
        "    sorted_bam,\n",
        "    dedup_bam\n",
        "])\n",
        "\n",
        "# Step 8: Generate alignment statistics\n",
        "print(\"Generating alignment statistics...\")\n",
        "subprocess.run(['samtools', 'flagstat', dedup_bam, '-o', f'{output_prefix}.flagstat.txt'])\n",
        "\n",
        "# Cleanup intermediate files (optional)\n",
        "print(\"Cleaning up intermediate files...\")\n",
        "os.remove(sam_file)\n",
        "os.remove(bam_file)\n",
        "\n",
        "print(\"Pipeline completed successfully!\")\n",
        "print(f\"Final BAM file: {dedup_bam}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 384
        },
        "id": "zxawPQAxJXat",
        "outputId": "d0f8bd83-2c65-4c50-cbbd-20bdc0043968"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Running FastQC...\n",
            "Running fastp for quality trimming...\n",
            "Aligning to reference genome...\n",
            "Converting SAM to BAM...\n",
            "Sorting BAM...\n",
            "Indexing BAM...\n",
            "Marking duplicates...\n",
            "Generating alignment statistics...\n",
            "Cleaning up intermediate files...\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "FileNotFoundError",
          "evalue": "[Errno 2] No such file or directory: 'data/ngs_ws_hd701.bam'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-4-4dce72f4a1af>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     73\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Cleaning up intermediate files...\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msam_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 75\u001b[0;31m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbam_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     76\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     77\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Pipeline completed successfully!\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/ngs_ws_hd701.bam'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget -O data/NA12878_chr17_R1.fastq.gz https://github.com/genome-in-a-bottle/giab_data_indexes/raw/master/NA12878/sequence/NIST7035_TAAGGCGA_L001_R1_001_trimmed_example.fastq.gz\n",
        "!wget -O data/NA12878_chr17_R2.fastq.gz https://github.com/genome-in-a-bottle/giab_data_indexes/raw/master/NA12878/sequence/NIST7035_TAAGGCGA_L001_R2_001_trimmed_example.fastq.gz\n",
        "\n",
        "# If the above links don't work, try these alternative small datasets\n",
        "# !wget -O data/NA12878_chr17_R1.fastq.gz https://zenodo.org/record/1251112/files/NA12878_chr17_mini_R1.fastq.gz\n",
        "# !wget -O data/NA12878_chr17_R2.fastq.gz https://zenodo.org/record/1251112/files/NA12878_chr17_mini_R2.fastq.gz\n",
        "\n",
        "# 1. Examine content of the FASTQ file\n",
        "!zcat data/NA12878_chr17_R1.fastq.gz | head -n 20\n",
        "\n",
        "# 2. Perform FASTQ processing prior to alignment\n",
        "!fastp -i data/NA12878_chr17_R1.fastq.gz -o data/NA12878_chr17_R1_processed.fq.gz -I data/NA12878_chr17_R2.fastq.gz -O data/NA12878_chr17_R2_processed.fq.gz -z 4 -w 2 -h data/NA12878_fq_qc.html\n",
        "\n",
        "# 3. Generate FASTQ quality control data\n",
        "!fastqc -o data -f fastq data/*.gz\n",
        "\n",
        "# 4. Display QC report in Colab\n",
        "from IPython.display import HTML\n",
        "with open('data/NA12878_fq_qc.html', 'r') as f:\n",
        "    html_content = f.read()\n",
        "HTML(html_content)\n",
        "\n",
        "# 5. Align sequence to GRCh37\n",
        "!bwa mem -M -v 1 -t 2 -R \"@RG\\\\tID:NA12878\\\\tSM:NA12878\\\\tPL:ILLUMINA\\\\tPI:150\\\\tCN:lab\" assets/gatk_b37/human_g1k_v37.fasta data/NA12878_chr17_R1_processed.fq.gz data/NA12878_chr17_R2_processed.fq.gz > data/NA12878_raw.sam\n",
        "\n",
        "# 6. Convert the SAM file into a BAM file\n",
        "!samtools view -Shu data/NA12878_raw.sam > data/NA12878_raw.bam\n",
        "\n",
        "# 7. Sort and index raw BAM file\n",
        "!sambamba sort -p -t 2 -o data/NA12878_sorted.bam data/NA12878_raw.bam\n",
        "\n",
        "# 8. Mark PCR duplicates in BAM file\n",
        "!sambamba markdup -t 2 -p data/NA12878_sorted.bam data/NA12878_dedup.bam\n",
        "\n",
        "# 9. Perform indel realignment\n",
        "!java -jar abra2-2.23.jar --in data/NA12878_dedup.bam --out data/NA12878_realigned.bam --ref assets/gatk_b37/human_g1k_v37.fasta --threads 2 --targets assets/ngs_workshop.bed --index --tmpdir /tmp/ > data/NA12878_abra.log\n",
        "\n",
        "# 10. Inspect aligned sequences (show first few lines)\n",
        "!samtools view -h data/NA12878_realigned.bam | head -n 20\n",
        "\n",
        "# 11. Call variant\n",
        "!samtools mpileup -BA -q 20 -Q 30 -d 4000 -l assets/ngs_workshop.bed -f assets/gatk_b37/human_g1k_v37.fasta data/NA12878_realigned.bam | java -Xmx4G -jar VarScan.v2.4.6.jar mpileup2vcf --min-coverage 8 --min-var-freq 0.05 --p-value 0.05 --min-avg-qual 30 --strand-filter 1 --output-vcf 1 --variants > data/NA12878_raw.vcf\n",
        "\n",
        "# 12. Compress and index VCF file\n",
        "!bgzip data/NA12878_raw.vcf\n",
        "!tabix -p vcf data/NA12878_raw.vcf.gz\n",
        "\n",
        "# 13. Inspect variants in a raw VCF file\n",
        "!zcat data/NA12878_raw.vcf.gz | head -n 20\n",
        "\n",
        "# 14. Annotate variants (requires ANNOVAR to be set up)\n",
        "!perl annovar/table_annovar.pl data/NA12878_raw.vcf.gz assets/humandb/ -out data/NA12878_annotated -buildver hg19 -remove -nastring . -otherinfo -vcfinput -thread 2 -maxgenethread 2 -protocol refGene,cytoBand,cosmic85,clinvar_20150330 -operation g,r,f,f\n",
        "\n",
        "# 15. Inspect annotated VCF file\n",
        "!head -n 20 data/NA12878_annotated.hg19_multianno.vcf\n",
        "\n",
        "# 16. Visualize BAM files with IGV.js\n",
        "!pip install igv-jupyter\n",
        "import igv\n",
        "viewer = igv.Browser({\n",
        "    \"genome\": \"hg19\",\n",
        "    \"locus\": \"chr17:7,571,720-7,590,868\"  # TP53 region\n",
        "})\n",
        "viewer.load_track({\n",
        "    \"name\": \"NA12878 Alignment\",\n",
        "    \"url\": \"data/NA12878_realigned.bam\",\n",
        "    \"indexURL\": \"data/NA12878_realigned.bam.bai\",\n",
        "    \"format\": \"bam\"\n",
        "})\n",
        "viewer\n",
        "\n",
        "# Download results to your local machine\n",
        "from google.colab import files\n",
        "!zip -r ngs_results.zip data/\n",
        "files.download('ngs_results.zip')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "G4d-08tqIHEL",
        "outputId": "22cf7e30-c773-43c6-cf9a-94af5b71fba9"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-03-15 23:27:32--  https://github.com/genome-in-a-bottle/giab_data_indexes/raw/master/NA12878/sequence/NIST7035_TAAGGCGA_L001_R1_001_trimmed_example.fastq.gz\n",
            "Resolving github.com (github.com)... 140.82.112.3\n",
            "Connecting to github.com (github.com)|140.82.112.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 404 Not Found\n",
            "2025-03-15 23:27:33 ERROR 404: Not Found.\n",
            "\n",
            "--2025-03-15 23:27:33--  https://github.com/genome-in-a-bottle/giab_data_indexes/raw/master/NA12878/sequence/NIST7035_TAAGGCGA_L001_R2_001_trimmed_example.fastq.gz\n",
            "Resolving github.com (github.com)... 140.82.112.3\n",
            "Connecting to github.com (github.com)|140.82.112.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 404 Not Found\n",
            "2025-03-15 23:27:33 ERROR 404: Not Found.\n",
            "\n",
            "\n",
            "gzip: data/NA12878_chr17_R1.fastq.gz: unexpected end of file\n",
            "ERROR: igzip: Error invalid gzip header found: data/NA12878_chr17_R1.fastq.gz\n",
            "application/gzip\n",
            "application/gzip\n",
            "application/gzip\n",
            "application/gzip\n",
            "Failed to process data/NA12878_chr17_R1.fastq.gz\n",
            "java.io.EOFException\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readUByte(GZIPInputStream.java:269)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readUShort(GZIPInputStream.java:259)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readHeader(GZIPInputStream.java:165)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:80)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:92)\n",
            "\tat uk.ac.babraham.FastQC.Utilities.MultiMemberGZIPInputStream.<init>(MultiMemberGZIPInputStream.java:37)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.FastQFile.<init>(FastQFile.java:84)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:87)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:62)\n",
            "\tat uk.ac.babraham.FastQC.Analysis.OfflineRunner.processFile(OfflineRunner.java:163)\n",
            "\tat uk.ac.babraham.FastQC.Analysis.OfflineRunner.<init>(OfflineRunner.java:125)\n",
            "\tat uk.ac.babraham.FastQC.FastQCApplication.main(FastQCApplication.java:316)\n",
            "application/gzip\n",
            "Failed to process data/NA12878_chr17_R2.fastq.gz\n",
            "java.io.EOFException\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readUByte(GZIPInputStream.java:269)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readUShort(GZIPInputStream.java:259)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readHeader(GZIPInputStream.java:165)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:80)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:92)\n",
            "\tat uk.ac.babraham.FastQC.Utilities.MultiMemberGZIPInputStream.<init>(MultiMemberGZIPInputStream.java:37)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.FastQFile.<init>(FastQFile.java:84)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:87)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:62)\n",
            "\tat uk.ac.babraham.FastQC.Analysis.OfflineRunner.processFile(OfflineRunner.java:163)\n",
            "\tat uk.ac.babraham.FastQC.Analysis.OfflineRunner.<init>(OfflineRunner.java:125)\n",
            "\tat uk.ac.babraham.FastQC.FastQCApplication.main(FastQCApplication.java:316)\n",
            "Started analysis of ERR3239276_1_processed.fq.gz\n",
            "Approx 5% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 10% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 15% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 20% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 25% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 30% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 35% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 40% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 45% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 50% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 55% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 60% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 65% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 70% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 75% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 80% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 85% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 90% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 95% complete for ERR3239276_1_processed.fq.gz\n",
            "Analysis complete for ERR3239276_1_processed.fq.gz\n",
            "Started analysis of ERR3239276_2.fastq.gz\n",
            "Approx 5% complete for ERR3239276_2.fastq.gz\n",
            "Approx 10% complete for ERR3239276_2.fastq.gz\n",
            "Approx 15% complete for ERR3239276_2.fastq.gz\n",
            "Approx 20% complete for ERR3239276_2.fastq.gz\n",
            "Approx 25% complete for ERR3239276_2.fastq.gz\n",
            "Approx 30% complete for ERR3239276_2.fastq.gz\n",
            "Approx 35% complete for ERR3239276_2.fastq.gz\n",
            "Approx 40% complete for ERR3239276_2.fastq.gz\n",
            "Approx 45% complete for ERR3239276_2.fastq.gz\n",
            "Approx 50% complete for ERR3239276_2.fastq.gz\n",
            "Approx 55% complete for ERR3239276_2.fastq.gz\n",
            "Approx 60% complete for ERR3239276_2.fastq.gz\n",
            "Approx 65% complete for ERR3239276_2.fastq.gz\n",
            "Approx 70% complete for ERR3239276_2.fastq.gz\n",
            "Approx 75% complete for ERR3239276_2.fastq.gz\n",
            "Approx 80% complete for ERR3239276_2.fastq.gz\n",
            "Approx 85% complete for ERR3239276_2.fastq.gz\n",
            "Approx 90% complete for ERR3239276_2.fastq.gz\n",
            "Approx 95% complete for ERR3239276_2.fastq.gz\n",
            "Failed to process file ERR3239276_2.fastq.gz\n",
            "uk.ac.babraham.FastQC.Sequence.SequenceFormatException: Ran out of data in the middle of a fastq entry.  Your file is probably truncated\n",
            "\tat uk.ac.babraham.FastQC.Sequence.FastQFile.readNext(FastQFile.java:187)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.FastQFile.next(FastQFile.java:129)\n",
            "\tat uk.ac.babraham.FastQC.Analysis.AnalysisRunner.run(AnalysisRunner.java:77)\n",
            "\tat java.base/java.lang.Thread.run(Thread.java:834)\n",
            "Started analysis of ERR3239276_2_processed.fq.gz\n",
            "Approx 5% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 10% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 15% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 20% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 25% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 30% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 35% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 40% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 45% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 50% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 55% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 60% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 65% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 70% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 75% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 80% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 85% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 90% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 95% complete for ERR3239276_2_processed.fq.gz\n",
            "Analysis complete for ERR3239276_2_processed.fq.gz\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "FileNotFoundError",
          "evalue": "[Errno 2] No such file or directory: 'data/NA12878_fq_qc.html'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-2-935083f18510>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;31m# 4. Display QC report in Colab\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mIPython\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdisplay\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mHTML\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'data/NA12878_fq_qc.html'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m     \u001b[0mhtml_content\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0mHTML\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhtml_content\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/NA12878_fq_qc.html'"
          ]
        }
      ]
    }
  ]
}