{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Eitan177/Bioinformatics_Bootcamp/blob/main/informatics_bootcamp.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Welcome to the GOAL Bioinformatics Bootcamp, 2024! To Setup this notebook please run the first three cells. This will download the packages you will need to participate"
      ],
      "metadata": {
        "id": "fKTuyMLINOKx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%writefile assets/ngs_workshop.bed\n",
        "chr17\t7571720\t7590868\tTP53\n",
        "chr3\t178866311\t178952152\tPIK3CA\n",
        "chr7\t55086714\t55324313\tEGFR"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DELC_7EwGfWt",
        "outputId": "0c25f266-b8c6-47d6-bd2f-d34144cde33f"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Writing assets/ngs_workshop.bed\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# NGS Bioinformatics Hands On Lab\n",
        "\n",
        "# Install required packages\n",
        "!apt-get update\n",
        "!apt-get install -y fastp fastqc bwa samtools tabix\n",
        "!pip install pysam\n",
        "\n",
        "# Optional: Install additional tools if not available in Colab\n",
        "!wget https://github.com/biod/sambamba/releases/download/v0.8.0/sambamba-0.8.0-linux-amd64-static.gz\n",
        "!gunzip sambamba-0.8.0-linux-amd64-static.gz\n",
        "!chmod +x sambamba-0.8.0-linux-amd64-static\n",
        "!mv sambamba-0.8.0-linux-amd64-static /usr/local/bin/sambamba\n",
        "\n",
        "# For ABRA2\n",
        "!wget https://github.com/mozack/abra2/releases/download/v2.23/abra2-2.23.jar\n",
        "!chmod +x abra2-2.23.jar\n",
        "\n",
        "# For VarScan\n",
        "!wget https://github.com/dkoboldt/varscan/raw/master/VarScan.v2.4.6.jar\n",
        "!chmod +x VarScan.v2.4.6.jar\n",
        "\n",
        "# for ANNOVAR\n",
        "!wget http://www.openbioinformatics.org/annovar/download/0wgxR2rIVP/annovar.latest.tar.gz\n",
        "!tar -zxvf annovar.latest.tar.gz\n",
        "!rm -r annovar.latest.tar.gz\n",
        "!rm -r annovar/humandb\n",
        "\n",
        "\n",
        "# Create directories for data and reference files\n",
        "!mkdir -p data assets/gatk_b37 assets/humandb annovar\n",
        "\n",
        "# Download pre-indexed reference genome\n",
        "!wget https://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/analysisSet/hg19.p13.plusMT.no_alt_analysis_set.bwa_index.tar.gz -O assets/hg19_bwa_index.tar.gz\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IkUOHAp-FmAu",
        "outputId": "653fb1ab-d713-4794-dd0a-fe6201e52489"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rGet:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]\n",
            "Hit:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease\n",
            "Hit:3 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Get:4 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\n",
            "Get:5 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\n",
            "Hit:7 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Hit:8 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Get:9 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]\n",
            "Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Get:11 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,677 kB]\n",
            "Get:12 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [8,762 kB]\n",
            "Get:13 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [2,692 kB]\n",
            "Get:14 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,237 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [3,000 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,535 kB]\n",
            "Fetched 20.3 MB in 5s (3,745 kB/s)\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  default-jre default-jre-headless fonts-dejavu-core fonts-dejavu-extra libapache-pom-java\n",
            "  libargs4j-java libatk-wrapper-java libatk-wrapper-java-jni libcommons-compress-java\n",
            "  libcommons-io-java libcommons-jexl2-java libcommons-lang3-java libcommons-logging-java\n",
            "  libcommons-math3-java libcommons-parent-java libfindbin-libs-perl libfontenc1 libhts3\n",
            "  libhtscodecs2 libhtsjdk-java libjbzip2-java libjson-simple-java libngs-java libngs-sdk-dev\n",
            "  libngs-sdk2 libsis-base-java libsis-base-jni libsis-jhdf5-java libsis-jhdf5-jni libsnappy-java\n",
            "  libsnappy-jni libxkbfile1 libxtst6 libxxf86dga1 openjdk-11-jre x11-utils\n",
            "Suggested packages:\n",
            "  libxz-java libcommons-io-java-doc libcommons-lang3-java-doc libavalon-framework-java\n",
            "  libcommons-logging-java-doc libexcalibur-logkit-java liblog4j1.2-java picard-tools\n",
            "  libjson-simple-doc cwltool mesa-utils\n",
            "The following NEW packages will be installed:\n",
            "  bwa default-jre default-jre-headless fastp fastqc fonts-dejavu-core fonts-dejavu-extra\n",
            "  libapache-pom-java libargs4j-java libatk-wrapper-java libatk-wrapper-java-jni\n",
            "  libcommons-compress-java libcommons-io-java libcommons-jexl2-java libcommons-lang3-java\n",
            "  libcommons-logging-java libcommons-math3-java libcommons-parent-java libfindbin-libs-perl\n",
            "  libfontenc1 libhts3 libhtscodecs2 libhtsjdk-java libjbzip2-java libjson-simple-java libngs-java\n",
            "  libngs-sdk-dev libngs-sdk2 libsis-base-java libsis-base-jni libsis-jhdf5-java libsis-jhdf5-jni\n",
            "  libsnappy-java libsnappy-jni libxkbfile1 libxtst6 libxxf86dga1 openjdk-11-jre samtools tabix\n",
            "  x11-utils\n",
            "0 upgraded, 41 newly installed, 0 to remove and 30 not upgraded.\n",
            "Need to get 13.5 MB of archives.\n",
            "After this operation, 29.4 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bwa amd64 0.7.17-6 [195 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/main amd64 default-jre-headless amd64 2:1.11-72build2 [3,042 B]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/main amd64 libxtst6 amd64 2:1.2.3-1build4 [13.4 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 openjdk-11-jre amd64 11.0.26+4-1ubuntu1~22.04 [214 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/main amd64 default-jre amd64 2:1.11-72build2 [896 B]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/universe amd64 fastp amd64 0.20.1+dfsg-1 [193 kB]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-math3-java all 3.6.1-3 [2,016 kB]\n",
            "Get:8 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libjbzip2-java all 0.9.1-7 [42.8 kB]\n",
            "Get:9 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libapache-pom-java all 18-1 [4,720 B]\n",
            "Get:10 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-parent-java all 43-1 [10.8 kB]\n",
            "Get:11 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-compress-java all 1.21-1 [575 kB]\n",
            "Get:12 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-logging-java all 1.2-2 [60.3 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-jexl2-java all 2.1.1-5 [251 kB]\n",
            "Get:14 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libjson-simple-java all 2.3.0-1 [61.0 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libngs-sdk2 amd64 2.11.2-3 [65.6 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libngs-sdk-dev amd64 2.11.2-3 [111 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libngs-java amd64 2.11.2-3 [75.8 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsnappy-jni amd64 1.1.8.3-1 [6,756 B]\n",
            "Get:19 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsnappy-java all 1.1.8.3-1 [93.9 kB]\n",
            "Get:20 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtsjdk-java all 2.24.1+dfsg-1 [1,644 kB]\n",
            "Get:21 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libfindbin-libs-perl all 2.190.02-1 [23.2 kB]\n",
            "Get:22 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libargs4j-java all 2.33-1.1 [137 kB]\n",
            "Get:23 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-io-java all 2.11.0-2 [297 kB]\n",
            "Get:24 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-lang3-java all 3.11-1 [526 kB]\n",
            "Get:25 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsis-base-jni amd64 18.09~pre1+git20180928.45fbd31+dfsg-3 [12.7 kB]\n",
            "Get:26 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsis-base-java all 18.09~pre1+git20180928.45fbd31+dfsg-3 [96.4 kB]\n",
            "Get:27 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsis-jhdf5-jni amd64 19.04.0+dfsg-4 [96.6 kB]\n",
            "Get:28 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsis-jhdf5-java all 19.04.0+dfsg-4 [1,490 kB]\n",
            "Get:29 http://archive.ubuntu.com/ubuntu jammy/universe amd64 fastqc all 0.11.9+dfsg-5 [370 kB]\n",
            "Get:30 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-dejavu-core all 2.37-2build1 [1,041 kB]\n",
            "Get:31 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-dejavu-extra all 2.37-2build1 [2,041 kB]\n",
            "Get:32 http://archive.ubuntu.com/ubuntu jammy/main amd64 libfontenc1 amd64 1:1.1.4-1build3 [14.7 kB]\n",
            "Get:33 http://archive.ubuntu.com/ubuntu jammy/main amd64 libxkbfile1 amd64 1:1.1.0-1build3 [71.8 kB]\n",
            "Get:34 http://archive.ubuntu.com/ubuntu jammy/main amd64 libxxf86dga1 amd64 2:1.1.5-0ubuntu3 [12.6 kB]\n",
            "Get:35 http://archive.ubuntu.com/ubuntu jammy/main amd64 x11-utils amd64 7.7+5build2 [206 kB]\n",
            "Get:36 http://archive.ubuntu.com/ubuntu jammy/main amd64 libatk-wrapper-java all 0.38.0-5build1 [53.1 kB]\n",
            "Get:37 http://archive.ubuntu.com/ubuntu jammy/main amd64 libatk-wrapper-java-jni amd64 0.38.0-5build1 [49.0 kB]\n",
            "Get:38 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]\n",
            "Get:39 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]\n",
            "Get:40 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]\n",
            "Get:41 http://archive.ubuntu.com/ubuntu jammy/universe amd64 tabix amd64 1.13+ds-2build1 [351 kB]\n",
            "Fetched 13.5 MB in 1s (18.5 MB/s)\n",
            "Extracting templates from packages: 100%\n",
            "Selecting previously unselected package bwa.\n",
            "(Reading database ... 126209 files and directories currently installed.)\n",
            "Preparing to unpack .../00-bwa_0.7.17-6_amd64.deb ...\n",
            "Unpacking bwa (0.7.17-6) ...\n",
            "Selecting previously unselected package default-jre-headless.\n",
            "Preparing to unpack .../01-default-jre-headless_2%3a1.11-72build2_amd64.deb ...\n",
            "Unpacking default-jre-headless (2:1.11-72build2) ...\n",
            "Selecting previously unselected package libxtst6:amd64.\n",
            "Preparing to unpack .../02-libxtst6_2%3a1.2.3-1build4_amd64.deb ...\n",
            "Unpacking libxtst6:amd64 (2:1.2.3-1build4) ...\n",
            "Selecting previously unselected package openjdk-11-jre:amd64.\n",
            "Preparing to unpack .../03-openjdk-11-jre_11.0.26+4-1ubuntu1~22.04_amd64.deb ...\n",
            "Unpacking openjdk-11-jre:amd64 (11.0.26+4-1ubuntu1~22.04) ...\n",
            "Selecting previously unselected package default-jre.\n",
            "Preparing to unpack .../04-default-jre_2%3a1.11-72build2_amd64.deb ...\n",
            "Unpacking default-jre (2:1.11-72build2) ...\n",
            "Selecting previously unselected package fastp.\n",
            "Preparing to unpack .../05-fastp_0.20.1+dfsg-1_amd64.deb ...\n",
            "Unpacking fastp (0.20.1+dfsg-1) ...\n",
            "Selecting previously unselected package libcommons-math3-java.\n",
            "Preparing to unpack .../06-libcommons-math3-java_3.6.1-3_all.deb ...\n",
            "Unpacking libcommons-math3-java (3.6.1-3) ...\n",
            "Selecting previously unselected package libjbzip2-java.\n",
            "Preparing to unpack .../07-libjbzip2-java_0.9.1-7_all.deb ...\n",
            "Unpacking libjbzip2-java (0.9.1-7) ...\n",
            "Selecting previously unselected package libapache-pom-java.\n",
            "Preparing to unpack .../08-libapache-pom-java_18-1_all.deb ...\n",
            "Unpacking libapache-pom-java (18-1) ...\n",
            "Selecting previously unselected package libcommons-parent-java.\n",
            "Preparing to unpack .../09-libcommons-parent-java_43-1_all.deb ...\n",
            "Unpacking libcommons-parent-java (43-1) ...\n",
            "Selecting previously unselected package libcommons-compress-java.\n",
            "Preparing to unpack .../10-libcommons-compress-java_1.21-1_all.deb ...\n",
            "Unpacking libcommons-compress-java (1.21-1) ...\n",
            "Selecting previously unselected package libcommons-logging-java.\n",
            "Preparing to unpack .../11-libcommons-logging-java_1.2-2_all.deb ...\n",
            "Unpacking libcommons-logging-java (1.2-2) ...\n",
            "Selecting previously unselected package libcommons-jexl2-java.\n",
            "Preparing to unpack .../12-libcommons-jexl2-java_2.1.1-5_all.deb ...\n",
            "Unpacking libcommons-jexl2-java (2.1.1-5) ...\n",
            "Selecting previously unselected package libjson-simple-java.\n",
            "Preparing to unpack .../13-libjson-simple-java_2.3.0-1_all.deb ...\n",
            "Unpacking libjson-simple-java (2.3.0-1) ...\n",
            "Selecting previously unselected package libngs-sdk2:amd64.\n",
            "Preparing to unpack .../14-libngs-sdk2_2.11.2-3_amd64.deb ...\n",
            "Unpacking libngs-sdk2:amd64 (2.11.2-3) ...\n",
            "Selecting previously unselected package libngs-sdk-dev:amd64.\n",
            "Preparing to unpack .../15-libngs-sdk-dev_2.11.2-3_amd64.deb ...\n",
            "Unpacking libngs-sdk-dev:amd64 (2.11.2-3) ...\n",
            "Selecting previously unselected package libngs-java:amd64.\n",
            "Preparing to unpack .../16-libngs-java_2.11.2-3_amd64.deb ...\n",
            "Unpacking libngs-java:amd64 (2.11.2-3) ...\n",
            "Selecting previously unselected package libsnappy-jni.\n",
            "Preparing to unpack .../17-libsnappy-jni_1.1.8.3-1_amd64.deb ...\n",
            "Unpacking libsnappy-jni (1.1.8.3-1) ...\n",
            "Selecting previously unselected package libsnappy-java.\n",
            "Preparing to unpack .../18-libsnappy-java_1.1.8.3-1_all.deb ...\n",
            "Unpacking libsnappy-java (1.1.8.3-1) ...\n",
            "Selecting previously unselected package libhtsjdk-java.\n",
            "Preparing to unpack .../19-libhtsjdk-java_2.24.1+dfsg-1_all.deb ...\n",
            "Unpacking libhtsjdk-java (2.24.1+dfsg-1) ...\n",
            "Selecting previously unselected package libfindbin-libs-perl.\n",
            "Preparing to unpack .../20-libfindbin-libs-perl_2.190.02-1_all.deb ...\n",
            "Unpacking libfindbin-libs-perl (2.190.02-1) ...\n",
            "Selecting previously unselected package libargs4j-java.\n",
            "Preparing to unpack .../21-libargs4j-java_2.33-1.1_all.deb ...\n",
            "Unpacking libargs4j-java (2.33-1.1) ...\n",
            "Selecting previously unselected package libcommons-io-java.\n",
            "Preparing to unpack .../22-libcommons-io-java_2.11.0-2_all.deb ...\n",
            "Unpacking libcommons-io-java (2.11.0-2) ...\n",
            "Selecting previously unselected package libcommons-lang3-java.\n",
            "Preparing to unpack .../23-libcommons-lang3-java_3.11-1_all.deb ...\n",
            "Unpacking libcommons-lang3-java (3.11-1) ...\n",
            "Selecting previously unselected package libsis-base-jni.\n",
            "Preparing to unpack .../24-libsis-base-jni_18.09~pre1+git20180928.45fbd31+dfsg-3_amd64.deb ...\n",
            "Unpacking libsis-base-jni (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Selecting previously unselected package libsis-base-java.\n",
            "Preparing to unpack .../25-libsis-base-java_18.09~pre1+git20180928.45fbd31+dfsg-3_all.deb ...\n",
            "Unpacking libsis-base-java (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Selecting previously unselected package libsis-jhdf5-jni.\n",
            "Preparing to unpack .../26-libsis-jhdf5-jni_19.04.0+dfsg-4_amd64.deb ...\n",
            "Unpacking libsis-jhdf5-jni (19.04.0+dfsg-4) ...\n",
            "Selecting previously unselected package libsis-jhdf5-java.\n",
            "Preparing to unpack .../27-libsis-jhdf5-java_19.04.0+dfsg-4_all.deb ...\n",
            "Unpacking libsis-jhdf5-java (19.04.0+dfsg-4) ...\n",
            "Selecting previously unselected package fastqc.\n",
            "Preparing to unpack .../28-fastqc_0.11.9+dfsg-5_all.deb ...\n",
            "Unpacking fastqc (0.11.9+dfsg-5) ...\n",
            "Selecting previously unselected package fonts-dejavu-core.\n",
            "Preparing to unpack .../29-fonts-dejavu-core_2.37-2build1_all.deb ...\n",
            "Unpacking fonts-dejavu-core (2.37-2build1) ...\n",
            "Selecting previously unselected package fonts-dejavu-extra.\n",
            "Preparing to unpack .../30-fonts-dejavu-extra_2.37-2build1_all.deb ...\n",
            "Unpacking fonts-dejavu-extra (2.37-2build1) ...\n",
            "Selecting previously unselected package libfontenc1:amd64.\n",
            "Preparing to unpack .../31-libfontenc1_1%3a1.1.4-1build3_amd64.deb ...\n",
            "Unpacking libfontenc1:amd64 (1:1.1.4-1build3) ...\n",
            "Selecting previously unselected package libxkbfile1:amd64.\n",
            "Preparing to unpack .../32-libxkbfile1_1%3a1.1.0-1build3_amd64.deb ...\n",
            "Unpacking libxkbfile1:amd64 (1:1.1.0-1build3) ...\n",
            "Selecting previously unselected package libxxf86dga1:amd64.\n",
            "Preparing to unpack .../33-libxxf86dga1_2%3a1.1.5-0ubuntu3_amd64.deb ...\n",
            "Unpacking libxxf86dga1:amd64 (2:1.1.5-0ubuntu3) ...\n",
            "Selecting previously unselected package x11-utils.\n",
            "Preparing to unpack .../34-x11-utils_7.7+5build2_amd64.deb ...\n",
            "Unpacking x11-utils (7.7+5build2) ...\n",
            "Selecting previously unselected package libatk-wrapper-java.\n",
            "Preparing to unpack .../35-libatk-wrapper-java_0.38.0-5build1_all.deb ...\n",
            "Unpacking libatk-wrapper-java (0.38.0-5build1) ...\n",
            "Selecting previously unselected package libatk-wrapper-java-jni:amd64.\n",
            "Preparing to unpack .../36-libatk-wrapper-java-jni_0.38.0-5build1_amd64.deb ...\n",
            "Unpacking libatk-wrapper-java-jni:amd64 (0.38.0-5build1) ...\n",
            "Selecting previously unselected package libhtscodecs2:amd64.\n",
            "Preparing to unpack .../37-libhtscodecs2_1.1.1-3_amd64.deb ...\n",
            "Unpacking libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Selecting previously unselected package libhts3:amd64.\n",
            "Preparing to unpack .../38-libhts3_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Selecting previously unselected package samtools.\n",
            "Preparing to unpack .../39-samtools_1.13-4_amd64.deb ...\n",
            "Unpacking samtools (1.13-4) ...\n",
            "Selecting previously unselected package tabix.\n",
            "Preparing to unpack .../40-tabix_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking tabix (1.13+ds-2build1) ...\n",
            "Setting up libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Setting up libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Setting up default-jre-headless (2:1.11-72build2) ...\n",
            "Setting up bwa (0.7.17-6) ...\n",
            "Setting up libxtst6:amd64 (2:1.2.3-1build4) ...\n",
            "Setting up libngs-sdk2:amd64 (2.11.2-3) ...\n",
            "Setting up libxxf86dga1:amd64 (2:1.1.5-0ubuntu3) ...\n",
            "Setting up openjdk-11-jre:amd64 (11.0.26+4-1ubuntu1~22.04) ...\n",
            "Setting up libfindbin-libs-perl (2.190.02-1) ...\n",
            "Setting up default-jre (2:1.11-72build2) ...\n",
            "Setting up libjson-simple-java (2.3.0-1) ...\n",
            "Setting up samtools (1.13-4) ...\n",
            "Setting up fastp (0.20.1+dfsg-1) ...\n",
            "Setting up libsis-base-jni (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Setting up libfontenc1:amd64 (1:1.1.4-1build3) ...\n",
            "Setting up libapache-pom-java (18-1) ...\n",
            "Setting up fonts-dejavu-core (2.37-2build1) ...\n",
            "Setting up libjbzip2-java (0.9.1-7) ...\n",
            "Setting up libngs-sdk-dev:amd64 (2.11.2-3) ...\n",
            "Setting up fonts-dejavu-extra (2.37-2build1) ...\n",
            "Setting up libcommons-math3-java (3.6.1-3) ...\n",
            "Setting up libsis-jhdf5-jni (19.04.0+dfsg-4) ...\n",
            "Setting up libxkbfile1:amd64 (1:1.1.0-1build3) ...\n",
            "Setting up libngs-java:amd64 (2.11.2-3) ...\n",
            "Setting up tabix (1.13+ds-2build1) ...\n",
            "Setting up libsnappy-jni (1.1.8.3-1) ...\n",
            "Setting up libargs4j-java (2.33-1.1) ...\n",
            "Setting up libcommons-parent-java (43-1) ...\n",
            "Setting up libcommons-logging-java (1.2-2) ...\n",
            "Setting up libcommons-jexl2-java (2.1.1-5) ...\n",
            "Setting up libcommons-lang3-java (3.11-1) ...\n",
            "Setting up x11-utils (7.7+5build2) ...\n",
            "Setting up libsnappy-java (1.1.8.3-1) ...\n",
            "Setting up libatk-wrapper-java (0.38.0-5build1) ...\n",
            "Setting up libcommons-compress-java (1.21-1) ...\n",
            "Setting up libcommons-io-java (2.11.0-2) ...\n",
            "Setting up libatk-wrapper-java-jni:amd64 (0.38.0-5build1) ...\n",
            "Setting up libsis-base-java (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Setting up libhtsjdk-java (2.24.1+dfsg-1) ...\n",
            "Setting up libsis-jhdf5-java (19.04.0+dfsg-4) ...\n",
            "Setting up fastqc (0.11.9+dfsg-5) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for mailcap (3.70+nmu1ubuntu1) ...\n",
            "Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "Collecting pysam\n",
            "  Downloading pysam-0.23.0-cp311-cp311-manylinux_2_28_x86_64.whl.metadata (1.6 kB)\n",
            "Downloading pysam-0.23.0-cp311-cp311-manylinux_2_28_x86_64.whl (26.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m26.0/26.0 MB\u001b[0m \u001b[31m14.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: pysam\n",
            "Successfully installed pysam-0.23.0\n",
            "--2025-03-20 19:59:57--  https://github.com/biod/sambamba/releases/download/v0.8.0/sambamba-0.8.0-linux-amd64-static.gz\n",
            "Resolving github.com (github.com)... 140.82.114.4\n",
            "Connecting to github.com (github.com)|140.82.114.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/4167482/7e411b00-3323-11eb-9622-b063c6aa7fc6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250320%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250320T195957Z&X-Amz-Expires=300&X-Amz-Signature=f8de38d496defb9f8ea1aacd36c94e27818d0db1032f71cb04312e059f370669&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dsambamba-0.8.0-linux-amd64-static.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2025-03-20 19:59:57--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/4167482/7e411b00-3323-11eb-9622-b063c6aa7fc6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250320%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250320T195957Z&X-Amz-Expires=300&X-Amz-Signature=f8de38d496defb9f8ea1aacd36c94e27818d0db1032f71cb04312e059f370669&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dsambamba-0.8.0-linux-amd64-static.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.110.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 4715787 (4.5M) [application/octet-stream]\n",
            "Saving to: ‘sambamba-0.8.0-linux-amd64-static.gz’\n",
            "\n",
            "sambamba-0.8.0-linu 100%[===================>]   4.50M  --.-KB/s    in 0.09s   \n",
            "\n",
            "2025-03-20 19:59:58 (51.0 MB/s) - ‘sambamba-0.8.0-linux-amd64-static.gz’ saved [4715787/4715787]\n",
            "\n",
            "--2025-03-20 19:59:58--  https://github.com/mozack/abra2/releases/download/v2.23/abra2-2.23.jar\n",
            "Resolving github.com (github.com)... 140.82.114.4\n",
            "Connecting to github.com (github.com)|140.82.114.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/66582878/e5e1b580-f6a6-11ea-89c8-33c01403f120?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250320%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250320T195958Z&X-Amz-Expires=300&X-Amz-Signature=3782d2cf27fabb314025994473e46f25ac708e27e6aae5bd71f64efa79196a48&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dabra2-2.23.jar&response-content-type=application%2Foctet-stream [following]\n",
            "--2025-03-20 19:59:58--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/66582878/e5e1b580-f6a6-11ea-89c8-33c01403f120?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250320%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250320T195958Z&X-Amz-Expires=300&X-Amz-Signature=3782d2cf27fabb314025994473e46f25ac708e27e6aae5bd71f64efa79196a48&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dabra2-2.23.jar&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 9025192 (8.6M) [application/octet-stream]\n",
            "Saving to: ‘abra2-2.23.jar’\n",
            "\n",
            "abra2-2.23.jar      100%[===================>]   8.61M  32.6MB/s    in 0.3s    \n",
            "\n",
            "2025-03-20 19:59:59 (32.6 MB/s) - ‘abra2-2.23.jar’ saved [9025192/9025192]\n",
            "\n",
            "--2025-03-20 19:59:59--  https://github.com/dkoboldt/varscan/raw/master/VarScan.v2.4.6.jar\n",
            "Resolving github.com (github.com)... 140.82.113.3\n",
            "Connecting to github.com (github.com)|140.82.113.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://raw.githubusercontent.com/dkoboldt/varscan/master/VarScan.v2.4.6.jar [following]\n",
            "--2025-03-20 19:59:59--  https://raw.githubusercontent.com/dkoboldt/varscan/master/VarScan.v2.4.6.jar\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 126611 (124K) [application/octet-stream]\n",
            "Saving to: ‘VarScan.v2.4.6.jar’\n",
            "\n",
            "VarScan.v2.4.6.jar  100%[===================>] 123.64K  --.-KB/s    in 0.03s   \n",
            "\n",
            "2025-03-20 19:59:59 (4.22 MB/s) - ‘VarScan.v2.4.6.jar’ saved [126611/126611]\n",
            "\n",
            "--2025-03-20 20:00:00--  http://www.openbioinformatics.org/annovar/download/0wgxR2rIVP/annovar.latest.tar.gz\n",
            "Resolving www.openbioinformatics.org (www.openbioinformatics.org)... 67.205.156.247\n",
            "Connecting to www.openbioinformatics.org (www.openbioinformatics.org)|67.205.156.247|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 280727629 (268M) [application/x-gzip]\n",
            "Saving to: ‘annovar.latest.tar.gz’\n",
            "\n",
            "annovar.latest.tar. 100%[===================>] 267.72M  42.9MB/s    in 5.9s    \n",
            "\n",
            "2025-03-20 20:00:06 (45.8 MB/s) - ‘annovar.latest.tar.gz’ saved [280727629/280727629]\n",
            "\n",
            "annovar/\n",
            "annovar/example/\n",
            "annovar/example/ex1.avinput\n",
            "annovar/example/example.simple_region\n",
            "annovar/example/example.tab_region\n",
            "annovar/example/ex2.vcf\n",
            "annovar/example/grantham.matrix\n",
            "annovar/example/snplist.txt\n",
            "annovar/example/README\n",
            "annovar/example/gene_xref.txt\n",
            "annovar/example/gene_fullxref.txt\n",
            "annovar/humandb/\n",
            "annovar/humandb/hg19_refGene.txt\n",
            "annovar/humandb/hg19_refGeneMrna.fa\n",
            "annovar/humandb/hg19_refGeneVersion.txt\n",
            "annovar/humandb/hg19_refGeneWithVer.txt\n",
            "annovar/humandb/hg19_refGeneWithVerMrna.fa\n",
            "annovar/humandb/hg38_ensGeneMrna.fa\n",
            "annovar/humandb/hg38_ensGene.txt\n",
            "annovar/humandb/hg38_refGeneWithVer.txt\n",
            "annovar/humandb/hg38_refGeneWithVerMrna.fa\n",
            "annovar/humandb/hg19_example_db_generic.txt\n",
            "annovar/humandb/hg19_example_db_gff3.txt\n",
            "annovar/humandb/GRCh37_MT_ensGene.txt\n",
            "annovar/humandb/GRCh37_MT_ensGeneMrna.fa\n",
            "annovar/humandb/hg19_MT_ensGene.txt\n",
            "annovar/humandb/hg19_MT_ensGeneMrna.fa\n",
            "annovar/humandb/genometrax-sample-files-gff/\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_mirna_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_microsatellites_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_hgmd_disease_genes_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_evs_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_pathway_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_tcga_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_vista_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_cpg_islands_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_transfac_sites_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_dbnsfp_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_omim_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/list\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_dnase_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_hgmd_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_dbscsnv_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_clinvar_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_fulldbsnp_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_complete_genomics_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_tss_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_ptms_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_orpha_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_ethnicsnp_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_chip_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_drug_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_hgmdimputed_featuretype_hg38.gff\n",
            "annovar/humandb/genometrax-sample-files-gff/sample_pgmd_featuretype_hg38.gff\n",
            "annovar/annotate_variation.pl\n",
            "annovar/table_annovar.pl\n",
            "annovar/convert2annovar.pl\n",
            "annovar/retrieve_seq_from_fasta.pl\n",
            "annovar/coding_change.pl\n",
            "annovar/variants_reduction.pl\n",
            "--2025-03-20 20:00:19--  https://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/analysisSet/hg19.p13.plusMT.no_alt_analysis_set.bwa_index.tar.gz\n",
            "Resolving hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)|128.114.119.163|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 3443961889 (3.2G) [application/x-gzip]\n",
            "Saving to: ‘assets/hg19_bwa_index.tar.gz’\n",
            "\n",
            "assets/hg19_bwa_ind 100%[===================>]   3.21G  68.6MB/s    in 68s     \n",
            "\n",
            "2025-03-20 20:01:28 (48.0 MB/s) - ‘assets/hg19_bwa_index.tar.gz’ saved [3443961889/3443961889]\n",
            "\n",
            "mv: cannot stat 'assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set.*': No such file or directory\n",
            "mv: cannot stat 'assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set.fa': No such file or directory\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/analysisSet/hg19.p13.plusMT.no_alt_analysis_set.fa.gz -O assets/hg19fa_index.tar.gz"
      ],
      "metadata": {
        "id": "PCLBrAfFSDo0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!tar -xzf assets/hg19_bwa_index.tar.gz -C assets/gatk_b37/\n",
        "\n",
        "!mv assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set.* assets/gatk_b37/\n",
        "!mv assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set.fa assets/gatk_b37/human_g1k_v37.fasta"
      ],
      "metadata": {
        "id": "ARgpZPa8RhAT",
        "outputId": "7953d6f4-bfc5-4896-e18c-edcc3f7d57ba",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "mv: cannot stat 'assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set.*': No such file or directory\n",
            "mv: cannot stat 'assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set.fa': No such file or directory\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/analysisSet/hg19.p13.plusMT.no_alt_analysis_set.fa.gz -O assets/hg19fa_index.gz\n",
        "!gunzip assets/hg19fa_index.gz -C assets/gatk_b37/"
      ],
      "metadata": {
        "id": "QM7Of6eXSS1g",
        "outputId": "366e6b73-b507-4518-e7f9-6ed477c291d7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-03-20 20:41:53--  https://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/analysisSet/hg19.p13.plusMT.no_alt_analysis_set.fa.gz\n",
            "Resolving hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)|128.114.119.163|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 863259336 (823M) [application/x-gzip]\n",
            "Saving to: ‘assets/hg19fa_index.gz’\n",
            "\n",
            "assets/hg19fa_index 100%[===================>] 823.27M  72.8MB/s    in 11s     \n",
            "\n",
            "2025-03-20 20:42:05 (71.8 MB/s) - ‘assets/hg19fa_index.gz’ saved [863259336/863259336]\n",
            "\n",
            "gzip: invalid option -- 'C'\n",
            "Try `gzip --help' for more information.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# prompt: decompress /content/assets/hg19fa_index.gz and put in  assets/gatk_b37/\n",
        "\n",
        "\n",
        "!gunzip /content/assets/hg19fa_index.gz -c > assets/gatk_b37/hg19.fa\n"
      ],
      "metadata": {
        "id": "wEA2jlqJStHZ"
      },
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# 1. Examine content of the FASTQ file\n",
        "!zcat /content/ngs_ws_hd701_R1.fq.gz | head -n 20\n",
        "\n",
        "# 2. Perform FASTQ processing prior to alignment\n",
        "!fastp -i /content/ngs_ws_hd701_R1.fq.gz -o data/hd701_R1_processed.fq.gz -I /content/ngs_ws_hd701_R2.fq.gz -O data/hd701_R2_processed.fq.gz -z 4 -w 2 -h data/hd701_fq_qc.html\n",
        "\n",
        "# 3. Generate FASTQ quality control data\n",
        "!fastqc -o data -f fastq /content/ngs_ws_hd701_R1.fq.gz /content/ngs_ws_hd701_R2.fq.gz\n",
        "\n",
        "# 4. Display QC report in Colab\n",
        "from IPython.display import HTML\n",
        "with open('data/hd701_fq_qc.html', 'r') as f:\n",
        "    html_content = f.read()\n",
        "HTML(html_content)\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4hZqzfN5J0r5",
        "outputId": "49a99648-b782-41aa-946b-90547997ab4e"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "@NB501446:28:HMG5HBGX2:1:11101:10612:17818/1\n",
            "GTCAATAATTCTGTTCTTAGGCCACTTTTCTCCCTTAATTTATTTCCTGTTCTTTCTCTTTTTAGCTCCCAGGGTT\n",
            "+\n",
            "AAAAAEEEEEEEEEEEEEEEEEEEEEEEEEEEEAAEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEE6EEEEEE\n",
            "@NB501446:28:HMG5HBGX2:1:11101:16871:6164/1\n",
            "GCCATTGCAGTGCAGAAGGGCAGTATCGGCAACCTCCTGGGCTTCCTAGAGGACAGCCTTGCCGCCACACAGTTCC\n",
            "+\n",
            "AAAAAEEEEEEEEEEAAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE/EE/E<EEEEEEEEEEE/EEAAAEAEE\n",
            "@NB501446:28:HMG5HBGX2:1:11101:17659:14137/1\n",
            "GTTTGCTGGCCTGTTGAAAAATGACTGTAACAAAAGTGCTTCTGGTTATTTAACAGATGAAAATGAAGTGGGGTTT\n",
            "+\n",
            "AAAAAEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEE\n",
            "@NB501446:28:HMG5HBGX2:1:11101:23347:13640/1\n",
            "ATATGTAAATGTAGTATAGGGAAGCTTCATAAGTCAGTCTCATCTGCAAATACTTGTGGGATTTTTAGCACAGCAA\n",
            "+\n",
            "AAAAAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE\n",
            "@NB501446:28:HMG5HBGX2:1:11101:7116:18158/1\n",
            "CCGCCTGGAGAAGTTGTATAGCACTATGGTGCGCTTCCTCAGTGACCGAAAGAACCCGGTGTGCCGGGAGATGGCT\n",
            "+\n",
            "AAAAAAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEAEEEEE\n",
            "Read1 before filtering:\n",
            "total reads: 5026\n",
            "total bases: 381943\n",
            "Q20 bases: 372022(97.4025%)\n",
            "Q30 bases: 367505(96.2199%)\n",
            "\n",
            "Read2 before filtering:\n",
            "total reads: 5026\n",
            "total bases: 381976\n",
            "Q20 bases: 371763(97.3263%)\n",
            "Q30 bases: 367322(96.1636%)\n",
            "\n",
            "Read1 after filtering:\n",
            "total reads: 4959\n",
            "total bases: 376830\n",
            "Q20 bases: 368496(97.7884%)\n",
            "Q30 bases: 364330(96.6829%)\n",
            "\n",
            "Read2 aftering filtering:\n",
            "total reads: 4959\n",
            "total bases: 376873\n",
            "Q20 bases: 368234(97.7077%)\n",
            "Q30 bases: 364087(96.6073%)\n",
            "\n",
            "Filtering result:\n",
            "reads passed filter: 9918\n",
            "reads failed due to low quality: 130\n",
            "reads failed due to too many N: 4\n",
            "reads failed due to too short: 0\n",
            "reads with adapter trimmed: 6\n",
            "bases trimmed due to adapters: 22\n",
            "\n",
            "Duplication rate: 7.81121%\n",
            "\n",
            "Insert size peak (evaluated by paired-end reads): 117\n",
            "\n",
            "JSON report: fastp.json\n",
            "HTML report: data/hd701_fq_qc.html\n",
            "\n",
            "fastp -i /content/ngs_ws_hd701_R1.fq.gz -o data/hd701_R1_processed.fq.gz -I /content/ngs_ws_hd701_R2.fq.gz -O data/hd701_R2_processed.fq.gz -z 4 -w 2 -h data/hd701_fq_qc.html \n",
            "fastp v0.20.1, time used: 1 seconds\n",
            "Started analysis of ngs_ws_hd701_R1.fq.gz\n",
            "Approx 20% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 40% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 60% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 80% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 100% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Analysis complete for ngs_ws_hd701_R1.fq.gz\n",
            "Started analysis of ngs_ws_hd701_R2.fq.gz\n",
            "Approx 20% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 40% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 60% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 80% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 100% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Analysis complete for ngs_ws_hd701_R2.fq.gz\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (144, 177, 224)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 384)\n",
            "[M::mem_pestat] mean and std.dev: (186.75, 58.21)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 464)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem -M -v 1 -t 2 -R @RG\\tID:hd701\\tSM:hd701\\tPL:ILLUMINA\\tPI:150\\tCN:lab /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.p13.plusMT.no_alt_analysis_set.fa data/hd701_R1_processed.fq.gz data/hd701_R2_processed.fq.gz\n",
            "[main] Real time: 31.863 sec; CPU: 7.941 sec\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 5. Align sequence to GRCh37 - Update the BWA reference path\n",
        "!bwa mem -M -v 1 -t 2 -R \"@RG\\\\tID:hd701\\\\tSM:hd701\\\\tPL:ILLUMINA\\\\tPI:150\\\\tCN:lab\" /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.p13.plusMT.no_alt_analysis_set.fa data/hd701_R1_processed.fq.gz data/hd701_R2_processed.fq.gz > data/hd701_raw.sam\n",
        "\n"
      ],
      "metadata": {
        "id": "DNiIZprhTEfh",
        "outputId": "3bdbe9b5-de6c-415c-96f7-a3ce8784f982",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (144, 177, 224)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 384)\n",
            "[M::mem_pestat] mean and std.dev: (186.75, 58.21)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 464)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem -M -v 1 -t 2 -R @RG\\tID:hd701\\tSM:hd701\\tPL:ILLUMINA\\tPI:150\\tCN:lab /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.p13.plusMT.no_alt_analysis_set.fa data/hd701_R1_processed.fq.gz data/hd701_R2_processed.fq.gz\n",
            "[main] Real time: 33.805 sec; CPU: 8.244 sec\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 6. Convert the SAM file into a BAM file\n",
        "!samtools view -Shu data/hd701_raw.sam > data/hd701_raw.bam\n",
        "\n",
        "# 7. Sort and index raw BAM file\n",
        "!sambamba sort -p -t 2 -o data/hd701_sorted.bam data/hd701_raw.bam\n",
        "\n",
        "# 8. Mark PCR duplicates in BAM file\n",
        "!sambamba markdup -t 2 -p data/hd701_sorted.bam data/hd701_dedup.bam"
      ],
      "metadata": {
        "id": "sHwuMotpQ0BN",
        "outputId": "2805d5c1-5945-4480-a6be-52471787cdf9",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "sambamba 0.8.0\n",
            " by Artem Tarasov and Pjotr Prins (C) 2012-2020\n",
            "    LDC 1.10.0 / DMD v2.080.1 / LLVM6.0.1 / bootstrap LDC - the LLVM D compiler (0.17.4)\n",
            "\n",
            "Writing sorted chunks to temporary directory...\n",
            "[==============================================================================]\n",
            "\n",
            "sambamba 0.8.0\n",
            " by Artem Tarasov and Pjotr Prins (C) 2012-2020\n",
            "    LDC 1.10.0 / DMD v2.080.1 / LLVM6.0.1 / bootstrap LDC - the LLVM D compiler (0.17.4)\n",
            "\n",
            "finding positions of the duplicate reads in the file...\n",
            "[==============================================================================]\n",
            "  sorted 4959 end pairs\n",
            "     and 0 single ends (among them 0 unmatched pairs)\n",
            "  collecting indices of duplicate reads...   done in 2 ms\n",
            "  found 860 duplicates\n",
            "collected list of positions in 0 min 0 sec\n",
            "marking duplicates...\n",
            "[==============================================================================]\n",
            "collected list of positions in 0 min 0 sec\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 9. Perform indel realignment - Update the reference path\n",
        "!java -jar abra2-2.23.jar --in data/hd701_dedup.bam --out data/hd701_realigned.bam --ref /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa --threads 2 --targets assets/ngs_ws.bed --index --tmpdir /tmp/ > data/hd701_abra.log\n",
        "\n",
        "# 10. Inspect aligned sequences (show first few lines)\n",
        "!samtools view -h data/hd701_realigned.bam | head -n 20\n",
        "\n"
      ],
      "metadata": {
        "id": "TPom2kN_QAYl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools index data/hd701_realigned.bam"
      ],
      "metadata": {
        "id": "6lhXsb1VVkjQ"
      },
      "execution_count": 27,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!ls -l /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa\n",
        "!head -n 5 /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa"
      ],
      "metadata": {
        "id": "k4mo8gDDVl5s",
        "outputId": "2fbaa8c6-f6b8-4ea9-ed3c-d8d2ba377253",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "-rw-r--r-- 1 root root 3146301083 Mar 20 20:43 /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa\n",
            ">chr1  AC:CM000663.1  gi:224384768  LN:249250621  rl:Chromosome  M5:d949d5ece55a7f11abb62abfe947815d  AS:GRCh37.p13\n",
            "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
            "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
            "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
            "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!ls -l assets/ngs_ws.bed\n",
        "!head -n 5 assets/ngs_ws.bed"
      ],
      "metadata": {
        "id": "j7v57zFSVq1K",
        "outputId": "2951644a-6d94-4839-cabd-a9a0c897084b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "-rw-r--r-- 1 root root 186 Mar 20 20:18 assets/ngs_ws.bed\n",
            "1\t27105337\t27107952\tARID1A\n",
            "3\t178951769\t178952287\tPIK3CA_ex21\n",
            "7\t55241584\t55241766\tEGFR_ex18\n",
            "7\t140453045\t140453223\tBRAF_ex15\n",
            "12\t25398208\t25398329\tKRAS_ex2\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!java -version\n",
        "!ls -l VarScan.v2.4.6.jar"
      ],
      "metadata": {
        "id": "QAqBlhNTVufX",
        "outputId": "3f4061c0-9a47-4c21-fdac-bfb23886613b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 33,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "openjdk version \"11.0.26\" 2025-01-21\n",
            "OpenJDK Runtime Environment (build 11.0.26+4-post-Ubuntu-1ubuntu122.04)\n",
            "OpenJDK 64-Bit Server VM (build 11.0.26+4-post-Ubuntu-1ubuntu122.04, mixed mode, sharing)\n",
            "-rwxr-xr-x 1 root root 126611 Mar 20 19:59 VarScan.v2.4.6.jar\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!head -n 10 data/hd701_mpileup.txt\n",
        "!wc -l data/hd701_mpileup.txt"
      ],
      "metadata": {
        "id": "EQN_IDoFWmHZ",
        "outputId": "851f6055-b389-4979-9bcb-0662b49a5bae",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 37,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "chr1\t27105338\tT\t8\t........\tEEEEEEEE\n",
            "chr1\t27105339\tG\t8\t........\tEEEEAEEE\n",
            "chr1\t27105340\tT\t9\t.$.......^].\tEEEEEEEEA\n",
            "chr1\t27105341\tT\t7\t.......\tEEEEEEA\n",
            "chr1\t27105342\tG\t7\t.......\tEAEEEEA\n",
            "chr1\t27105343\tG\t8\t........\tEEEEEEEA\n",
            "chr1\t27105344\tG\t8\t........\tEEEEEEEA\n",
            "chr1\t27105345\tC\t8\t........\tEEEAEEEE\n",
            "chr1\t27105346\tA\t8\t........\tAEEEEEEE\n",
            "chr1\t27105347\tT\t8\t........\tEEEEEEEE\n",
            "8078 data/hd701_mpileup.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools mpileup -BA -q 20 -Q 30 -d 4000 -l assets/ngs_ws.bed -f /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa data/hd701_realigned.bam > data/hd701_mpileup.txt\n",
        "!wc -l data/hd701_mpileup.txt\n",
        "# Step 2: Run VarScan on the mpileup file\n",
        "#!cat data/hd701_mpileup.txt | java -Xmx4G -jar VarScan.v2.4.6.jar mpileup2vcf --min-coverage 8 --min-var-freq 0.05 --p-value 0.05 --min-avg-qual 30 --strand-filter 1 --output-vcf 1 --variants > data/hd701_raw.vcf"
      ],
      "metadata": {
        "id": "nZHuJyLiV3I4",
        "outputId": "87fb36d1-7adf-41cb-ad32-b065b0805357",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 36,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[mpileup] 1 samples in 1 input files\n",
            "8078 data/hd701_mpileup.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!# 11. Call variant - Update the reference path\n",
        "!samtools mpileup -BA -q 20 -Q 30 -d 4000 -l assets/ngs_ws.bed -f /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa data/hd701_realigned.bam | java -Xmx4G -jar VarScan.v2.4.6.jar mpileup2vcf --min-coverage 8 --min-var-freq 0.05 --p-value 0.05 --min-avg-qual 30 --strand-filter 1 --output-vcf 1 --variants > data/hd701_raw.vcf"
      ],
      "metadata": {
        "id": "vSBVTyONUxOF",
        "outputId": "74d3213b-be53-45af-d9e6-9a9b8dece57f",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 38,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[mpileup] 1 samples in 1 input files\n",
            "Only variants will be reported\n",
            "Min coverage:\t8\n",
            "Min reads2:\t2\n",
            "Min var freq:\t0.05\n",
            "Min avg qual:\t30\n",
            "P-value thresh:\t0.05\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Reading input from STDIN\n",
            "8078 bases in pileup file\n",
            "14 variant positions (10 SNP, 4 indel)\n",
            "0 were failed by the strand-filter\n",
            "14 variant positions reported (10 SNP, 4 indel)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 12. Compress and index VCF file\n",
        "!bgzip data/hd701_raw.vcf\n",
        "!tabix -p vcf data/hd701_raw.vcf.gz\n",
        "\n",
        "# 13. Inspect variants in a raw VCF file\n",
        "!zcat data/hd701_raw.vcf.gz | head -n 20\n",
        "\n"
      ],
      "metadata": {
        "id": "10pXsS4ZXjUB",
        "outputId": "cca3a977-2a5c-481d-bd75-0f9d6faed360",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 39,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bgzip] data/hd701_raw.vcf.gz already exists; do you wish to overwrite (y or n)? y\n",
            "##fileformat=VCFv4.3\n",
            "##source=VarScan2\n",
            "##INFO=<ID=ADP,Number=1,Type=Integer,Description=\"Average per-sample depth of bases with Phred score >= 30\">\n",
            "##INFO=<ID=WT,Number=1,Type=Integer,Description=\"Number of samples called reference (wild-type)\">\n",
            "##INFO=<ID=HET,Number=1,Type=Integer,Description=\"Number of samples called heterozygous-variant\">\n",
            "##INFO=<ID=HOM,Number=1,Type=Integer,Description=\"Number of samples called homozygous-variant\">\n",
            "##INFO=<ID=NC,Number=1,Type=Integer,Description=\"Number of samples not called\">\n",
            "##FILTER=<ID=str10,Description=\"Less than 10% or more than 90% of variant supporting reads on one strand\">\n",
            "##FILTER=<ID=indelError,Description=\"Likely artifact due to indel reads at this position\">\n",
            "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
            "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
            "##FORMAT=<ID=SDP,Number=1,Type=Integer,Description=\"Raw Read Depth as reported by SAMtools\">\n",
            "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Quality Read Depth of bases with Phred score >= 30\">\n",
            "##FORMAT=<ID=RD,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases (reads1)\">\n",
            "##FORMAT=<ID=AD,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases (reads2)\">\n",
            "##FORMAT=<ID=FREQ,Number=1,Type=String,Description=\"Variant allele frequency\">\n",
            "##FORMAT=<ID=PVAL,Number=1,Type=String,Description=\"P-value from Fisher's Exact Test\">\n",
            "##FORMAT=<ID=RBQ,Number=1,Type=Integer,Description=\"Average quality of reference-supporting bases (qual1)\">\n",
            "##FORMAT=<ID=ABQ,Number=1,Type=Integer,Description=\"Average quality of variant-supporting bases (qual2)\">\n",
            "##FORMAT=<ID=RDF,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on forward strand (reads1plus)\">\n",
            "NOTICE: the --polish argument is set ON automatically (use --nopolish to change this behavior)\n",
            "Error: the required database file assets/humandb/hg19_refGene.txt does not exist.\n",
            "head: cannot open 'data/hd701_annotated.hg19_multianno.vcf' for reading: No such file or directory\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!perl annovar/annotate_variation.pl -buildver hg19 -downdb -webfrom annovar refGene assets/humandb/\n",
        "!perl annovar/annotate_variation.pl -buildver hg19 -downdb -webfrom annovar cytoBand assets/humandb/\n",
        "!perl annovar/annotate_variation.pl -buildver hg19 -downdb -webfrom annovar clinvar_20150330 assets/humandb/\n",
        "!perl annovar/annotate_variation.pl -buildver hg19 -downdb -webfrom annovar cosmic85 assets/humandb/"
      ],
      "metadata": {
        "id": "96htTEtgX969",
        "outputId": "3ceceb98-d48d-4a1b-f7a9-e4f04e89d08c",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 41,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_refGene.txt.gz ... OK\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_refGeneMrna.fa.gz ... OK\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_refGeneVersion.txt.gz ... OK\n",
            "NOTICE: Uncompressing downloaded files\n",
            "NOTICE: Finished downloading annotation files for hg19 build version, with files saved at the 'assets/humandb' directory\n",
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.gz ... Failed\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.idx.gz ... Failed\n",
            "WARNING: Some files cannot be downloaded, including http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.idx.gz, http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.gz\n",
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_clinvar_20150330.txt.gz ... OK\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_clinvar_20150330.txt.idx.gz ... OK\n",
            "NOTICE: Uncompressing downloaded files\n",
            "NOTICE: Finished downloading annotation files for hg19 build version, with files saved at the 'assets/humandb' directory\n",
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_cosmic85.txt.gz ... Failed\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_cosmic85.txt.idx.gz ... Failed\n",
            "WARNING: Some files cannot be downloaded, including http://www.openbioinformatics.org/annovar/download/hg19_cosmic85.txt.gz, http://www.openbioinformatics.org/annovar/download/hg19_cosmic85.txt.idx.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!perl /content/annovar/annotate_variation.pl -buildver hg19 -downdb cytoBand assets/humandb/\n"
      ],
      "metadata": {
        "id": "6dxKn1fxYOIe",
        "outputId": "d680940e-2bc7-4ed0-dd2a-0f9333a01a47",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 64,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/cytoBand.txt.gz ... OK\n",
            "NOTICE: Uncompressing downloaded files\n",
            "NOTICE: Finished downloading annotation files for hg19 build version, with files saved at the 'assets/humandb' directory\n",
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/hg19_cosmic85.txt.txt.gz ... Failed\n",
            "WARNING: Some files cannot be downloaded, including http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/hg19_cosmic85.txt.txt.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!perl /content/annovar/annotate_variation.pl -buildver hg19 -downdb hg19_cosmic assets/humandb/"
      ],
      "metadata": {
        "id": "5cpJuPPdcOLK",
        "outputId": "2179dd5d-a12d-4954-981b-1014fd973993",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 67,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/hg19_cosmic.txt.gz ... Failed\n",
            "WARNING: Some files cannot be downloaded, including http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/hg19_cosmic.txt.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "annotate_variation.pl -filter -build hg19 -out ex4 -dbtype cosmic68 ex4.avinput humandb/"
      ],
      "metadata": {
        "id": "tVXz0a1NdB7c"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Gw3ifCzBYy0k",
        "outputId": "2361216f-c285-4c3d-f587-86405865cb39",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 51,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "NOTICE: Web-based checking to see whether ANNOVAR new version is available ... Done\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.gz ... Failed\n",
            "NOTICE: Downloading annotation database http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.idx.gz ... Failed\n",
            "WARNING: Some files cannot be downloaded, including http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.gz, http://www.openbioinformatics.org/annovar/download/hg19_cytoBand.txt.idx.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 14. Annotate variants (requires ANNOVAR to be set up)\n",
        "!perl annovar/table_annovar.pl data/hd701_raw.vcf.gz assets/humandb/ -out data/hd701_annotated -buildver hg19 -remove -nastring . -otherinfo -vcfinput -thread 2 -maxgenethread 2 -protocol refGene,cytoBand,cosmic85,clinvar_20150330 -operation g,r,f,f\n",
        "\n",
        "# 15. Inspect annotated VCF file\n",
        "!head -n 20 data/hd701_annotated.hg19_multianno.vcf"
      ],
      "metadata": {
        "id": "MeRm3HGFXqbe",
        "outputId": "91a8b835-1fd8-49d1-c811-52585b6c4aca",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 62,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "NOTICE: the --polish argument is set ON automatically (use --nopolish to change this behavior)\n",
            "Error: the required database file assets/humandb/hg19_cosmic85.txt does not exist.\n",
            "head: cannot open 'data/hd701_annotated.hg19_multianno.vcf' for reading: No such file or directory\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 11. Call variant - Update the reference path\n",
        "!samtools mpileup -BA -q 20 -Q 30 -d 4000 -l assets/ngs_ws.bed -f /content/assets/gatk_b37/hg19.p13.plusMT.no_alt_analysis_set/hg19.fa data/hd701_realigned.bam | java -Xmx4G -jar VarScan.v2.4.6.jar mpileup2vcf --min-coverage 8 --min-var-freq 0.05 --p-value 0.05 --min-avg-qual 30 --strand-filter 1 --output-vcf 1 --variants > data/hd701_raw.vcf\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "# 16. Visualize BAM files with IGV.js\n",
        "!pip install igv-jupyter\n",
        "import igv\n",
        "viewer = igv.Browser({\n",
        "    \"genome\": \"hg19\",\n",
        "    \"locus\": \"chr17:7,571,720-7,590,868\"  # TP53 region\n",
        "})\n",
        "viewer.load_track({\n",
        "    \"name\": \"HD701 Alignment\",\n",
        "    \"url\": \"data/hd701_realigned.bam\",\n",
        "    \"indexURL\": \"data/hd701_realigned.bam.bai\",\n",
        "    \"format\": \"bam\"\n",
        "})\n",
        "viewer\n",
        "\n",
        "# Download results to your local machine\n",
        "from google.colab import files\n",
        "!zip -r hd701_results.zip data/\n",
        "files.download('hd701_results.zip')"
      ],
      "metadata": {
        "id": "6aB5zjpZQSsI",
        "outputId": "bd94a03c-834c-430d-b200-0e6f64ac91f4",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Only variants will be reported\n",
            "Min coverage:\t8\n",
            "Min reads2:\t2\n",
            "Min var freq:\t0.05\n",
            "Min avg qual:\t30\n",
            "P-value thresh:\t0.05\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "Input stream not ready, waiting for 5 seconds...\n",
            "[bgzip] data/hd701_raw.vcf.gz already exists; do you wish to overwrite (y or n)? ^C\n",
            "[tabix] the index file exists. Please use '-f' to overwrite.\n",
            "NOTICE: the --polish argument is set ON automatically (use --nopolish to change this behavior)\n",
            "Error: the required database file assets/humandb/hg19_refGene.txt does not exist.\n",
            "head: cannot open 'data/hd701_annotated.hg19_multianno.vcf' for reading: No such file or directory\n",
            "Requirement already satisfied: igv-jupyter in /usr/local/lib/python3.11/dist-packages (2.0.1)\n",
            "Requirement already satisfied: jupyterlab~=3.0 in /usr/local/lib/python3.11/dist-packages (from igv-jupyter) (3.6.8)\n",
            "Requirement already satisfied: ipywidgets>=7.0.0 in /usr/local/lib/python3.11/dist-packages (from igv-jupyter) (7.7.1)\n",
            "Requirement already satisfied: igv-notebook in /usr/local/lib/python3.11/dist-packages (from igv-jupyter) (0.6.2)\n",
            "Requirement already satisfied: ipykernel>=4.5.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (6.17.1)\n",
            "Requirement already satisfied: ipython-genutils~=0.2.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (0.2.0)\n",
            "Requirement already satisfied: traitlets>=4.3.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (5.7.1)\n",
            "Requirement already satisfied: widgetsnbextension~=3.6.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (3.6.10)\n",
            "Requirement already satisfied: ipython>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (7.34.0)\n",
            "Requirement already satisfied: jupyterlab-widgets>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets>=7.0.0->igv-jupyter) (3.0.13)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (24.2)\n",
            "Requirement already satisfied: tornado>=6.1.0 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (6.4.2)\n",
            "Requirement already satisfied: jupyter-core in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (5.7.2)\n",
            "Requirement already satisfied: jupyterlab-server~=2.19 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (2.27.3)\n",
            "Requirement already satisfied: jupyter-server<3,>=1.16.0 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (2.15.0)\n",
            "Requirement already satisfied: jupyter-ydoc~=0.2.4 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (0.2.5)\n",
            "Requirement already satisfied: jupyter-server-ydoc~=0.8.0 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (0.8.0)\n",
            "Requirement already satisfied: nbclassic in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (1.2.0)\n",
            "Requirement already satisfied: notebook<7 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (6.5.7)\n",
            "Requirement already satisfied: jinja2>=2.1 in /usr/local/lib/python3.11/dist-packages (from jupyterlab~=3.0->igv-jupyter) (3.1.6)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from igv-notebook->igv-jupyter) (2.32.3)\n",
            "Requirement already satisfied: debugpy>=1.0 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (1.8.0)\n",
            "Requirement already satisfied: jupyter-client>=6.1.12 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (7.4.9)\n",
            "Requirement already satisfied: matplotlib-inline>=0.1 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (0.1.7)\n",
            "Requirement already satisfied: nest-asyncio in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (1.6.0)\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (5.9.5)\n",
            "Requirement already satisfied: pyzmq>=17 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (24.0.1)\n",
            "Requirement already satisfied: setuptools>=18.5 in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (75.1.0)\n",
            "Requirement already satisfied: jedi>=0.16 in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.19.2)\n",
            "Requirement already satisfied: decorator in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (4.4.2)\n",
            "Requirement already satisfied: pickleshare in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.7.5)\n",
            "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (3.0.50)\n",
            "Requirement already satisfied: pygments in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (2.18.0)\n",
            "Requirement already satisfied: backcall in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.2.0)\n",
            "Requirement already satisfied: pexpect>4.3 in /usr/local/lib/python3.11/dist-packages (from ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (4.9.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2>=2.1->jupyterlab~=3.0->igv-jupyter) (3.0.2)\n",
            "Requirement already satisfied: anyio>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (4.9.0)\n",
            "Requirement already satisfied: argon2-cffi>=21.1 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (23.1.0)\n",
            "Requirement already satisfied: jupyter-events>=0.11.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.12.0)\n",
            "Requirement already satisfied: jupyter-server-terminals>=0.4.4 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.5.3)\n",
            "Requirement already satisfied: nbconvert>=6.4.4 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (7.16.6)\n",
            "Requirement already satisfied: nbformat>=5.3.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (5.10.4)\n",
            "Requirement already satisfied: overrides>=5.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (7.7.0)\n",
            "Requirement already satisfied: prometheus-client>=0.9 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.21.1)\n",
            "Requirement already satisfied: send2trash>=1.8.2 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.8.3)\n",
            "Requirement already satisfied: terminado>=0.8.3 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.18.1)\n",
            "Requirement already satisfied: websocket-client>=1.7 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.8.0)\n",
            "Requirement already satisfied: platformdirs>=2.5 in /usr/local/lib/python3.11/dist-packages (from jupyter-core->jupyterlab~=3.0->igv-jupyter) (4.3.6)\n",
            "Requirement already satisfied: jupyter-server-fileid<1,>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server-ydoc~=0.8.0->jupyterlab~=3.0->igv-jupyter) (0.9.3)\n",
            "Requirement already satisfied: ypy-websocket<0.9.0,>=0.8.2 in /usr/local/lib/python3.11/dist-packages (from jupyter-server-ydoc~=0.8.0->jupyterlab~=3.0->igv-jupyter) (0.8.4)\n",
            "Requirement already satisfied: y-py<0.7.0,>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-ydoc~=0.2.4->jupyterlab~=3.0->igv-jupyter) (0.6.2)\n",
            "Requirement already satisfied: babel>=2.10 in /usr/local/lib/python3.11/dist-packages (from jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (2.17.0)\n",
            "Requirement already satisfied: json5>=0.9.0 in /usr/local/lib/python3.11/dist-packages (from jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (0.10.0)\n",
            "Requirement already satisfied: jsonschema>=4.18.0 in /usr/local/lib/python3.11/dist-packages (from jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (4.23.0)\n",
            "Requirement already satisfied: notebook-shim>=0.2.3 in /usr/local/lib/python3.11/dist-packages (from nbclassic->jupyterlab~=3.0->igv-jupyter) (0.2.4)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->igv-notebook->igv-jupyter) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests->igv-notebook->igv-jupyter) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->igv-notebook->igv-jupyter) (2.3.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests->igv-notebook->igv-jupyter) (2025.1.31)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.11/dist-packages (from anyio>=3.1.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.3.1)\n",
            "Requirement already satisfied: typing_extensions>=4.5 in /usr/local/lib/python3.11/dist-packages (from anyio>=3.1.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (4.12.2)\n",
            "Requirement already satisfied: argon2-cffi-bindings in /usr/local/lib/python3.11/dist-packages (from argon2-cffi>=21.1->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (21.2.0)\n",
            "Requirement already satisfied: parso<0.9.0,>=0.8.4 in /usr/local/lib/python3.11/dist-packages (from jedi>=0.16->ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.8.4)\n",
            "Requirement already satisfied: attrs>=22.2.0 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (25.3.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (2024.10.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (0.36.2)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab~=3.0->igv-jupyter) (0.23.1)\n",
            "Requirement already satisfied: entrypoints in /usr/local/lib/python3.11/dist-packages (from jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (0.4)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (2.8.2)\n",
            "Requirement already satisfied: python-json-logger>=2.0.4 in /usr/local/lib/python3.11/dist-packages (from jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (3.3.0)\n",
            "Requirement already satisfied: pyyaml>=5.3 in /usr/local/lib/python3.11/dist-packages (from jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (6.0.2)\n",
            "Requirement already satisfied: rfc3339-validator in /usr/local/lib/python3.11/dist-packages (from jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.1.4)\n",
            "Requirement already satisfied: rfc3986-validator>=0.1.1 in /usr/local/lib/python3.11/dist-packages (from jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.1.1)\n",
            "Requirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (4.13.3)\n",
            "Requirement already satisfied: bleach!=5.0.0 in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (6.2.0)\n",
            "Requirement already satisfied: defusedxml in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.7.1)\n",
            "Requirement already satisfied: jupyterlab-pygments in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.3.0)\n",
            "Requirement already satisfied: mistune<4,>=2.0.3 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (3.1.2)\n",
            "Requirement already satisfied: nbclient>=0.5.0 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.10.2)\n",
            "Requirement already satisfied: pandocfilters>=1.4.1 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.5.1)\n",
            "Requirement already satisfied: fastjsonschema>=2.15 in /usr/local/lib/python3.11/dist-packages (from nbformat>=5.3.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (2.21.1)\n",
            "Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.11/dist-packages (from pexpect>4.3->ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.7.0)\n",
            "Requirement already satisfied: wcwidth in /usr/local/lib/python3.11/dist-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=4.0.0->ipywidgets>=7.0.0->igv-jupyter) (0.2.13)\n",
            "Requirement already satisfied: aiofiles<23,>=22.1.0 in /usr/local/lib/python3.11/dist-packages (from ypy-websocket<0.9.0,>=0.8.2->jupyter-server-ydoc~=0.8.0->jupyterlab~=3.0->igv-jupyter) (22.1.0)\n",
            "Requirement already satisfied: aiosqlite<1,>=0.17.0 in /usr/local/lib/python3.11/dist-packages (from ypy-websocket<0.9.0,>=0.8.2->jupyter-server-ydoc~=0.8.0->jupyterlab~=3.0->igv-jupyter) (0.21.0)\n",
            "Requirement already satisfied: webencodings in /usr/local/lib/python3.11/dist-packages (from bleach!=5.0.0->bleach[css]!=5.0.0->nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (0.5.1)\n",
            "Requirement already satisfied: tinycss2<1.5,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.4.0)\n",
            "Requirement already satisfied: fqdn in /usr/local/lib/python3.11/dist-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.5.1)\n",
            "Requirement already satisfied: isoduration in /usr/local/lib/python3.11/dist-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (20.11.0)\n",
            "Requirement already satisfied: jsonpointer>1.13 in /usr/local/lib/python3.11/dist-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (3.0.0)\n",
            "Requirement already satisfied: uri-template in /usr/local/lib/python3.11/dist-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.3.0)\n",
            "Requirement already satisfied: webcolors>=24.6.0 in /usr/local/lib/python3.11/dist-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (24.11.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets>=7.0.0->igv-jupyter) (1.17.0)\n",
            "Requirement already satisfied: cffi>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from argon2-cffi-bindings->argon2-cffi>=21.1->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.17.1)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4->nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (2.6)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.11/dist-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi>=21.1->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (2.22)\n",
            "Requirement already satisfied: arrow>=0.15.0 in /usr/local/lib/python3.11/dist-packages (from isoduration->jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (1.3.0)\n",
            "Requirement already satisfied: types-python-dateutil>=2.8.10 in /usr/local/lib/python3.11/dist-packages (from arrow>=0.15.0->isoduration->jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.11.0->jupyter-server<3,>=1.16.0->jupyterlab~=3.0->igv-jupyter) (2.9.0.20241206)\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "AttributeError",
          "evalue": "module 'igv' has no attribute 'Browser'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-24-76971d50519f>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'pip install igv-jupyter'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0migv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m viewer = igv.Browser({\n\u001b[0m\u001b[1;32m     22\u001b[0m     \u001b[0;34m\"genome\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m\"hg19\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m     \u001b[0;34m\"locus\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m\"chr17:7,571,720-7,590,868\"\u001b[0m  \u001b[0;31m# TP53 region\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mAttributeError\u001b[0m: module 'igv' has no attribute 'Browser'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!fastqc -o data -f fastq /content/ngs_ws_hd701_R1.fq.gz /content/ngs_ws_hd701_R2.fq.gz"
      ],
      "metadata": {
        "id": "Kuvbnln8OA_n",
        "outputId": "15c35715-2af8-4210-d15f-571e54a0c646",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of ngs_ws_hd701_R1.fq.gz\n",
            "Approx 20% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 40% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 60% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 80% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Approx 100% complete for ngs_ws_hd701_R1.fq.gz\n",
            "Analysis complete for ngs_ws_hd701_R1.fq.gz\n",
            "Started analysis of ngs_ws_hd701_R2.fq.gz\n",
            "Approx 20% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 40% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 60% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 80% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Approx 100% complete for ngs_ws_hd701_R2.fq.gz\n",
            "Analysis complete for ngs_ws_hd701_R2.fq.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import subprocess\n",
        "\n",
        "# Define input and output paths\n",
        "fastq_r1 = '/content/ngs_ws_hd701_R1.fq.gz'\n",
        "fastq_r2 = '/content/ngs_ws_hd701_R2.fq.gz'\n",
        "ref_genome = 'assets/gatk_b37/human_g1k_v37.fasta'\n",
        "output_prefix = 'data/ngs_ws_hd701'\n",
        "threads = 4  # Adjust based on your environment\n",
        "\n",
        "# Step 1: Quality control with FastQC (optional but recommended)\n",
        "print(\"Running FastQC...\")\n",
        "os.makedirs('data/fastqc', exist_ok=True)\n",
        "subprocess.run(['fastqc', fastq_r1, fastq_r2, '-o', 'data/fastqc'])\n",
        "\n",
        "# Step 2: Quality trimming with fastp\n",
        "print(\"Running fastp for quality trimming...\")\n",
        "subprocess.run([\n",
        "    'fastp',\n",
        "    '-i', fastq_r1,\n",
        "    '-I', fastq_r2,\n",
        "    '-o', f'{output_prefix}_trimmed_R1.fastq.gz',\n",
        "    '-O', f'{output_prefix}_trimmed_R2.fastq.gz',\n",
        "    '--json', f'{output_prefix}_fastp.json',\n",
        "    '--html', f'{output_prefix}_fastp.html'\n",
        "])\n",
        "\n",
        "# Step 3: Align to reference with BWA-MEM\n",
        "print(\"Aligning to reference genome...\")\n",
        "sam_file = f'{output_prefix}.sam'\n",
        "bwa_cmd = [\n",
        "    'bwa', 'mem',\n",
        "    '-t', str(threads),\n",
        "    '-R', f'@RG\\\\tID:ngs_ws_hd701\\\\tSM:ngs_ws_hd701\\\\tPL:ILLUMINA',\n",
        "    ref_genome,\n",
        "    f'{output_prefix}_trimmed_R1.fastq.gz',\n",
        "    f'{output_prefix}_trimmed_R2.fastq.gz'\n",
        "]\n",
        "\n",
        "with open(sam_file, 'w') as f:\n",
        "    subprocess.run(bwa_cmd, stdout=f)\n",
        "\n",
        "# Step 4: Convert SAM to BAM\n",
        "print(\"Converting SAM to BAM...\")\n",
        "bam_file = f'{output_prefix}.bam'\n",
        "subprocess.run(['samtools', 'view', '-b', '-h', '-o', bam_file, sam_file])\n",
        "\n",
        "# Step 5: Sort the BAM file\n",
        "print(\"Sorting BAM...\")\n",
        "sorted_bam = f'{output_prefix}.sorted.bam'\n",
        "subprocess.run(['samtools', 'sort', '-o', sorted_bam, bam_file])\n",
        "\n",
        "# Step 6: Index the BAM file\n",
        "print(\"Indexing BAM...\")\n",
        "subprocess.run(['samtools', 'index', sorted_bam])\n",
        "\n",
        "# Step 7: Mark duplicates with sambamba\n",
        "print(\"Marking duplicates...\")\n",
        "dedup_bam = f'{output_prefix}.dedup.bam'\n",
        "subprocess.run([\n",
        "    'sambamba', 'markdup',\n",
        "    '--tmpdir=/tmp',\n",
        "    '-t', str(threads),\n",
        "    sorted_bam,\n",
        "    dedup_bam\n",
        "])\n",
        "\n",
        "# Step 8: Generate alignment statistics\n",
        "print(\"Generating alignment statistics...\")\n",
        "subprocess.run(['samtools', 'flagstat', dedup_bam, '-o', f'{output_prefix}.flagstat.txt'])\n",
        "\n",
        "# Cleanup intermediate files (optional)\n",
        "print(\"Cleaning up intermediate files...\")\n",
        "os.remove(sam_file)\n",
        "os.remove(bam_file)\n",
        "\n",
        "print(\"Pipeline completed successfully!\")\n",
        "print(f\"Final BAM file: {dedup_bam}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 384
        },
        "id": "zxawPQAxJXat",
        "outputId": "d0f8bd83-2c65-4c50-cbbd-20bdc0043968"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Running FastQC...\n",
            "Running fastp for quality trimming...\n",
            "Aligning to reference genome...\n",
            "Converting SAM to BAM...\n",
            "Sorting BAM...\n",
            "Indexing BAM...\n",
            "Marking duplicates...\n",
            "Generating alignment statistics...\n",
            "Cleaning up intermediate files...\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "FileNotFoundError",
          "evalue": "[Errno 2] No such file or directory: 'data/ngs_ws_hd701.bam'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-4-4dce72f4a1af>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     73\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Cleaning up intermediate files...\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msam_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 75\u001b[0;31m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbam_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     76\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     77\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Pipeline completed successfully!\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/ngs_ws_hd701.bam'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget -O data/NA12878_chr17_R1.fastq.gz https://github.com/genome-in-a-bottle/giab_data_indexes/raw/master/NA12878/sequence/NIST7035_TAAGGCGA_L001_R1_001_trimmed_example.fastq.gz\n",
        "!wget -O data/NA12878_chr17_R2.fastq.gz https://github.com/genome-in-a-bottle/giab_data_indexes/raw/master/NA12878/sequence/NIST7035_TAAGGCGA_L001_R2_001_trimmed_example.fastq.gz\n",
        "\n",
        "# If the above links don't work, try these alternative small datasets\n",
        "# !wget -O data/NA12878_chr17_R1.fastq.gz https://zenodo.org/record/1251112/files/NA12878_chr17_mini_R1.fastq.gz\n",
        "# !wget -O data/NA12878_chr17_R2.fastq.gz https://zenodo.org/record/1251112/files/NA12878_chr17_mini_R2.fastq.gz\n",
        "\n",
        "# 1. Examine content of the FASTQ file\n",
        "!zcat data/NA12878_chr17_R1.fastq.gz | head -n 20\n",
        "\n",
        "# 2. Perform FASTQ processing prior to alignment\n",
        "!fastp -i data/NA12878_chr17_R1.fastq.gz -o data/NA12878_chr17_R1_processed.fq.gz -I data/NA12878_chr17_R2.fastq.gz -O data/NA12878_chr17_R2_processed.fq.gz -z 4 -w 2 -h data/NA12878_fq_qc.html\n",
        "\n",
        "# 3. Generate FASTQ quality control data\n",
        "!fastqc -o data -f fastq data/*.gz\n",
        "\n",
        "# 4. Display QC report in Colab\n",
        "from IPython.display import HTML\n",
        "with open('data/NA12878_fq_qc.html', 'r') as f:\n",
        "    html_content = f.read()\n",
        "HTML(html_content)\n",
        "\n",
        "# 5. Align sequence to GRCh37\n",
        "!bwa mem -M -v 1 -t 2 -R \"@RG\\\\tID:NA12878\\\\tSM:NA12878\\\\tPL:ILLUMINA\\\\tPI:150\\\\tCN:lab\" assets/gatk_b37/human_g1k_v37.fasta data/NA12878_chr17_R1_processed.fq.gz data/NA12878_chr17_R2_processed.fq.gz > data/NA12878_raw.sam\n",
        "\n",
        "# 6. Convert the SAM file into a BAM file\n",
        "!samtools view -Shu data/NA12878_raw.sam > data/NA12878_raw.bam\n",
        "\n",
        "# 7. Sort and index raw BAM file\n",
        "!sambamba sort -p -t 2 -o data/NA12878_sorted.bam data/NA12878_raw.bam\n",
        "\n",
        "# 8. Mark PCR duplicates in BAM file\n",
        "!sambamba markdup -t 2 -p data/NA12878_sorted.bam data/NA12878_dedup.bam\n",
        "\n",
        "# 9. Perform indel realignment\n",
        "!java -jar abra2-2.23.jar --in data/NA12878_dedup.bam --out data/NA12878_realigned.bam --ref assets/gatk_b37/human_g1k_v37.fasta --threads 2 --targets assets/ngs_workshop.bed --index --tmpdir /tmp/ > data/NA12878_abra.log\n",
        "\n",
        "# 10. Inspect aligned sequences (show first few lines)\n",
        "!samtools view -h data/NA12878_realigned.bam | head -n 20\n",
        "\n",
        "# 11. Call variant\n",
        "!samtools mpileup -BA -q 20 -Q 30 -d 4000 -l assets/ngs_workshop.bed -f assets/gatk_b37/human_g1k_v37.fasta data/NA12878_realigned.bam | java -Xmx4G -jar VarScan.v2.4.6.jar mpileup2vcf --min-coverage 8 --min-var-freq 0.05 --p-value 0.05 --min-avg-qual 30 --strand-filter 1 --output-vcf 1 --variants > data/NA12878_raw.vcf\n",
        "\n",
        "# 12. Compress and index VCF file\n",
        "!bgzip data/NA12878_raw.vcf\n",
        "!tabix -p vcf data/NA12878_raw.vcf.gz\n",
        "\n",
        "# 13. Inspect variants in a raw VCF file\n",
        "!zcat data/NA12878_raw.vcf.gz | head -n 20\n",
        "\n",
        "# 14. Annotate variants (requires ANNOVAR to be set up)\n",
        "!perl annovar/table_annovar.pl data/NA12878_raw.vcf.gz assets/humandb/ -out data/NA12878_annotated -buildver hg19 -remove -nastring . -otherinfo -vcfinput -thread 2 -maxgenethread 2 -protocol refGene,cytoBand,cosmic85,clinvar_20150330 -operation g,r,f,f\n",
        "\n",
        "# 15. Inspect annotated VCF file\n",
        "!head -n 20 data/NA12878_annotated.hg19_multianno.vcf\n",
        "\n",
        "# 16. Visualize BAM files with IGV.js\n",
        "!pip install igv-jupyter\n",
        "import igv\n",
        "viewer = igv.Browser({\n",
        "    \"genome\": \"hg19\",\n",
        "    \"locus\": \"chr17:7,571,720-7,590,868\"  # TP53 region\n",
        "})\n",
        "viewer.load_track({\n",
        "    \"name\": \"NA12878 Alignment\",\n",
        "    \"url\": \"data/NA12878_realigned.bam\",\n",
        "    \"indexURL\": \"data/NA12878_realigned.bam.bai\",\n",
        "    \"format\": \"bam\"\n",
        "})\n",
        "viewer\n",
        "\n",
        "# Download results to your local machine\n",
        "from google.colab import files\n",
        "!zip -r ngs_results.zip data/\n",
        "files.download('ngs_results.zip')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "G4d-08tqIHEL",
        "outputId": "22cf7e30-c773-43c6-cf9a-94af5b71fba9"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-03-15 23:27:32--  https://github.com/genome-in-a-bottle/giab_data_indexes/raw/master/NA12878/sequence/NIST7035_TAAGGCGA_L001_R1_001_trimmed_example.fastq.gz\n",
            "Resolving github.com (github.com)... 140.82.112.3\n",
            "Connecting to github.com (github.com)|140.82.112.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 404 Not Found\n",
            "2025-03-15 23:27:33 ERROR 404: Not Found.\n",
            "\n",
            "--2025-03-15 23:27:33--  https://github.com/genome-in-a-bottle/giab_data_indexes/raw/master/NA12878/sequence/NIST7035_TAAGGCGA_L001_R2_001_trimmed_example.fastq.gz\n",
            "Resolving github.com (github.com)... 140.82.112.3\n",
            "Connecting to github.com (github.com)|140.82.112.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 404 Not Found\n",
            "2025-03-15 23:27:33 ERROR 404: Not Found.\n",
            "\n",
            "\n",
            "gzip: data/NA12878_chr17_R1.fastq.gz: unexpected end of file\n",
            "ERROR: igzip: Error invalid gzip header found: data/NA12878_chr17_R1.fastq.gz\n",
            "application/gzip\n",
            "application/gzip\n",
            "application/gzip\n",
            "application/gzip\n",
            "Failed to process data/NA12878_chr17_R1.fastq.gz\n",
            "java.io.EOFException\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readUByte(GZIPInputStream.java:269)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readUShort(GZIPInputStream.java:259)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readHeader(GZIPInputStream.java:165)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:80)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:92)\n",
            "\tat uk.ac.babraham.FastQC.Utilities.MultiMemberGZIPInputStream.<init>(MultiMemberGZIPInputStream.java:37)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.FastQFile.<init>(FastQFile.java:84)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:87)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:62)\n",
            "\tat uk.ac.babraham.FastQC.Analysis.OfflineRunner.processFile(OfflineRunner.java:163)\n",
            "\tat uk.ac.babraham.FastQC.Analysis.OfflineRunner.<init>(OfflineRunner.java:125)\n",
            "\tat uk.ac.babraham.FastQC.FastQCApplication.main(FastQCApplication.java:316)\n",
            "application/gzip\n",
            "Failed to process data/NA12878_chr17_R2.fastq.gz\n",
            "java.io.EOFException\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readUByte(GZIPInputStream.java:269)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readUShort(GZIPInputStream.java:259)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.readHeader(GZIPInputStream.java:165)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:80)\n",
            "\tat java.base/java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:92)\n",
            "\tat uk.ac.babraham.FastQC.Utilities.MultiMemberGZIPInputStream.<init>(MultiMemberGZIPInputStream.java:37)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.FastQFile.<init>(FastQFile.java:84)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:87)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:62)\n",
            "\tat uk.ac.babraham.FastQC.Analysis.OfflineRunner.processFile(OfflineRunner.java:163)\n",
            "\tat uk.ac.babraham.FastQC.Analysis.OfflineRunner.<init>(OfflineRunner.java:125)\n",
            "\tat uk.ac.babraham.FastQC.FastQCApplication.main(FastQCApplication.java:316)\n",
            "Started analysis of ERR3239276_1_processed.fq.gz\n",
            "Approx 5% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 10% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 15% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 20% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 25% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 30% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 35% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 40% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 45% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 50% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 55% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 60% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 65% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 70% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 75% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 80% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 85% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 90% complete for ERR3239276_1_processed.fq.gz\n",
            "Approx 95% complete for ERR3239276_1_processed.fq.gz\n",
            "Analysis complete for ERR3239276_1_processed.fq.gz\n",
            "Started analysis of ERR3239276_2.fastq.gz\n",
            "Approx 5% complete for ERR3239276_2.fastq.gz\n",
            "Approx 10% complete for ERR3239276_2.fastq.gz\n",
            "Approx 15% complete for ERR3239276_2.fastq.gz\n",
            "Approx 20% complete for ERR3239276_2.fastq.gz\n",
            "Approx 25% complete for ERR3239276_2.fastq.gz\n",
            "Approx 30% complete for ERR3239276_2.fastq.gz\n",
            "Approx 35% complete for ERR3239276_2.fastq.gz\n",
            "Approx 40% complete for ERR3239276_2.fastq.gz\n",
            "Approx 45% complete for ERR3239276_2.fastq.gz\n",
            "Approx 50% complete for ERR3239276_2.fastq.gz\n",
            "Approx 55% complete for ERR3239276_2.fastq.gz\n",
            "Approx 60% complete for ERR3239276_2.fastq.gz\n",
            "Approx 65% complete for ERR3239276_2.fastq.gz\n",
            "Approx 70% complete for ERR3239276_2.fastq.gz\n",
            "Approx 75% complete for ERR3239276_2.fastq.gz\n",
            "Approx 80% complete for ERR3239276_2.fastq.gz\n",
            "Approx 85% complete for ERR3239276_2.fastq.gz\n",
            "Approx 90% complete for ERR3239276_2.fastq.gz\n",
            "Approx 95% complete for ERR3239276_2.fastq.gz\n",
            "Failed to process file ERR3239276_2.fastq.gz\n",
            "uk.ac.babraham.FastQC.Sequence.SequenceFormatException: Ran out of data in the middle of a fastq entry.  Your file is probably truncated\n",
            "\tat uk.ac.babraham.FastQC.Sequence.FastQFile.readNext(FastQFile.java:187)\n",
            "\tat uk.ac.babraham.FastQC.Sequence.FastQFile.next(FastQFile.java:129)\n",
            "\tat uk.ac.babraham.FastQC.Analysis.AnalysisRunner.run(AnalysisRunner.java:77)\n",
            "\tat java.base/java.lang.Thread.run(Thread.java:834)\n",
            "Started analysis of ERR3239276_2_processed.fq.gz\n",
            "Approx 5% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 10% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 15% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 20% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 25% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 30% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 35% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 40% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 45% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 50% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 55% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 60% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 65% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 70% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 75% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 80% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 85% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 90% complete for ERR3239276_2_processed.fq.gz\n",
            "Approx 95% complete for ERR3239276_2_processed.fq.gz\n",
            "Analysis complete for ERR3239276_2_processed.fq.gz\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "FileNotFoundError",
          "evalue": "[Errno 2] No such file or directory: 'data/NA12878_fq_qc.html'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-2-935083f18510>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;31m# 4. Display QC report in Colab\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mIPython\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdisplay\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mHTML\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'data/NA12878_fq_qc.html'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m     \u001b[0mhtml_content\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0mHTML\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhtml_content\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/NA12878_fq_qc.html'"
          ]
        }
      ]
    }
  ]
}